diff -ur busybox-0.60.3/init.c busybox-0.60.3.pseries/init.c
--- busybox-0.60.3/init.c	Fri Apr 26 23:12:55 2002
+++ busybox-0.60.3.pseries/init.c	Thu Oct 10 09:19:23 2002
@@ -295,7 +295,7 @@
 
 /* How much memory does this machine have?
    Units are kBytes to avoid overflow on 4GB machines */
-static int check_free_memory(void)
+static unsigned int check_free_memory(void)
 {
 	struct sysinfo info;
 	unsigned int result, u, s=10;
 
