--- busybox-1.9.1/scripts/trylink.orig	2008-03-10 21:50:00.000000000 +0100
+++ busybox-1.9.1/scripts/trylink	2008-03-10 21:59:17.000000000 +0100
@@ -47,11 +47,12 @@ try() {
 }
 
 check_cc() {
-    if $CC $1 -shared -o /dev/null -xc /dev/null > /dev/null 2>&1; then
+    if $CC $1 -shared -o /tmp/$$ -xc /dev/null > /dev/null 2>&1; then
         echo "$1";
     else
         echo "$2";
     fi
+    rm -f /tmp/$$ > /dev/null 2>&1
 }
 
 EXE="$1"
