--- busybox-1.1.1/include/platform.h.orig	2006-03-22 22:16:24.000000000 +0100
+++ busybox-1.1.1/include/platform.h	2006-08-02 17:47:22.000000000 +0200
@@ -85,11 +85,14 @@
 #endif
 
 #ifdef __BIG_ENDIAN__
-    #define BB_BIG_ENDIAN 1
+	#define BB_BIG_ENDIAN 1
+	#define BB_LITTLE_ENDIAN 0
 #elif __BYTE_ORDER == __BIG_ENDIAN
-    #define BB_BIG_ENDIAN 1
+	#define BB_BIG_ENDIAN 1
+	#define BB_LITTLE_ENDIAN 0
 #else
-    #define BB_BIG_ENDIAN 0
+	#define BB_BIG_ENDIAN 0
+	#define BB_LITTLE_ENDIAN 1
 #endif
 
 /* ---- Networking ------------------------------------------ */
