diff -nru devfsd/compat_name.c devfsd.new/compat_name.c
--- devfsd/compat_name.c	Fri Jan 25 02:06:35 2002
+++ devfsd.new/compat_name.c	Mon Sep  9 14:26:04 2002
@@ -187,6 +187,33 @@
 	    compat_name = write_old_sd_name (buffer, major, minor, ptr + 4);
 	return (compat_name);
     }
+    else if (strncmp (devname, "hd/disc", 7) == 0)
+    {   /* iSeries IDE devices (don't ask me why this is their root -SLD-) */
+        if (strncmp (ptr = (strrchr (devname, '/') + 1), "mt", 2) == 0)
+	{
+	    sprintf (buffer, "%sht%d", ptr + 2, minor & 0x7f);
+	    compat_name = buffer;
+	}
+	else if (strcmp (devname + namelen - 4, "disc") == 0)
+	{
+	    sprintf ( buffer, "hd%c",
+		      get_old_ide_name (major, minor) );
+	    compat_name = buffer;
+	}
+	else if (strncmp (ptr = (strrchr (devname, '/') + 1), "part", 4) == 0)
+	{
+	    sprintf (buffer, "hd%c%s",
+		     get_old_ide_name (major, minor), ptr + 4);
+	    compat_name = buffer;
+	}
+	else if (strcmp (devname + namelen - 2, "cd") == 0)
+	{
+	    sprintf ( buffer, "hd%c",
+		      get_old_ide_name (major, minor) );
+	    compat_name = buffer;
+	}
+	return (compat_name);
+    }
     else if (strncmp (devname, "ide/host", 8) == 0)
     {   /*  All IDE devices  */
 	if (strncmp (ptr = (strrchr (devname, '/') + 1), "mt", 2) == 0)
