diff -urN dhcp-2.0pl5.orig/common/tr.c dhcp-2.0pl5/common/tr.c
--- dhcp-2.0pl5.orig/common/tr.c	1999-06-10 02:58:06.000000000 +0200
+++ dhcp-2.0pl5/common/tr.c	2006-03-01 09:43:54.000000000 +0100
@@ -13,6 +13,8 @@
 #include "netinet/if_tr.h"
 #include <sys/time.h>
 
+typedef unsigned short __u16;
+
 /*
  * token ring device handling subroutines.  These are required as token-ring
  * does not have a simple on-the-wire header but requires the use of
