--- udev-079.orig/debian/rules.d/20-names.rules	2006-05-21 16:10:16.000000000 +0200
+++ udev-079/debian/rules.d/20-names.rules	2006-05-21 16:10:19.000000000 +0200
@@ -14,6 +14,26 @@
 KERNEL=="device-mapper",		NAME="mapper/control"
 KERNEL=="dm-[0-9]*",			NAME="dm/%n"
 
+# RAID controllers
+KERNEL=="cciss!*|ida!*|rd!*", PROGRAM="/lib/udev/raid-devfs.sh %k", \
+                                        NAME="%c{1}", SYMLINK+="%k"
+# other block devices
+KERNEL=="md[0-9]*",                     NAME="md/%n",
+KERNEL=="loop[0-9]*",                   NAME="loop/%n"
+KERNEL=="ram[0-9]*",                    NAME="rd/%n"
+
+# by-id (hardware serial number)
+KERNEL=="hd*[!0-9]", IMPORT{program}="/sbin/ata_id --export $tempnode"
+KERNEL=="hd*[!0-9]", ENV{ID_SERIAL}=="?*", SYMLINK+="disk/by-id/ata-$env{ID_MODEL}_$env{ID_SERIAL}"
+KERNEL=="hd*[0-9]", IMPORT{parent}=="ID_*", SYMLINK+="disk/by-id/ata-$env{ID_MODEL}_$env{ID_SERIAL}-part%n"
+
+KERNEL=="sd*[!0-9]|sr*", SYSFS{ieee1394_id}=="*", ENV{ID_SERIAL}="$sysfs{ieee1394_id}", ENV{ID_BUS}="ieee1394"
+KERNEL=="sd*[!0-9]|sr*", ENV{ID_SERIAL}=="", IMPORT{program}="/sbin/usb_id -x"
+KERNEL=="sd*[!0-9]|sr*", ENV{ID_SERIAL}=="", IMPORT{program}="/sbin/scsi_id -g -x -s %p"
+KERNEL=="sd*[!0-9]|sr*", ENV{ID_SERIAL}=="", IMPORT{program}="/sbin/scsi_id -g -x -a -s %p"
+KERNEL=="dasd*[!0-9]", IMPORT{program}="/sbin/dasd_id --export $tempnode"
+KERNEL=="sd*[!0-9]|sr*|dasd*[!0-9]", ENV{ID_SERIAL}=="?*", SYMLINK+="disk/by-id/$env{ID_BUS}-$env{ID_SERIAL}"
+
 # Input devices, group under /dev/input
 KERNEL=="event[0-9]*",			NAME="input/%k"
 KERNEL=="mice",				NAME="input/%k"
