--- udev-108.orig/debian/udev.init	2007-04-29 11:48:56.000000000 +0200
+++ udev-108/debian/udev.init	2007-04-29 11:54:37.000000000 +0200
@@ -40,7 +40,8 @@ case "$1" in
 	
 	# Start udevd
 	log_begin_msg "Starting kernel event manager..."
-	if start-stop-daemon --start --quiet --exec /sbin/udevd -- --daemon; then
+	#if start-stop-daemon --start --quiet --exec /sbin/udevd -- --daemon; then
+	if /sbin/udevd --daemon; then
 	    log_end_msg 0
 	else
 	    log_end_msg $?
