--- udev-079.orig/debian/rules	2006-03-24 20:41:49.000000000 -0800
+++ udev-079/debian/rules	2006-03-24 20:43:38.000000000 -0800
@@ -26,7 +26,7 @@
 # Apply patches to the package
 patch: patch-stamp
 patch-stamp:
-	dh_testdir
+	#dh_testdir
 	@patches=debian/patches/*.patch; for patch in $$patches; do \
 		test -f $$patch || continue; \
 		echo "Applying $$patch"; \
@@ -36,7 +36,7 @@
 
 # Remove patches from the package
 unpatch:
-	dh_testdir
+	#dh_testdir
 	@if test -f patch-stamp; then \
 		patches=debian/patches/*.patch; \
 		for patch in $$patches; do \
@@ -54,12 +54,12 @@
 # Build the package
 build: patch-stamp build-stamp
 build-stamp:
-	dh_testdir
+	#dh_testdir
 
 	test -d udeb || mkdir udeb
 	find -print0 | cpio -pdm0l udeb
 
-	$(MAKE) V=true USE_SELINUX=true EXTRAS="$(EXTRAS)" \
+	$(MAKE) V=true USE_SELINUX=false EXTRAS="$(EXTRAS)" \
 		OPTFLAGS="$(OPTFLAGS)" STRIPCMD=
 
 	$(MAKE) -C udeb V=true USE_SELINUX=false EXTRAS="$(EXTRAS)" \
@@ -70,11 +70,11 @@
 # Install files into the deb
 install-deb: DH_OPTIONS=-pudev
 install-deb: build
-	dh_testdir
-	dh_testroot
-	dh_clean -k 
-	dh_installdirs
-	dh_install
+	#dh_testdir
+	#dh_testroot
+	#dh_clean -k 
+	#dh_installdirs
+	#dh_install
 
 	$(MAKE) install V=true EXTRAS="$(EXTRAS)" DESTDIR=$(CURDIR)/debian/udev
 
@@ -137,17 +137,20 @@
 # Install files into the udeb
 install-udeb: DH_OPTIONS=-pudev-udeb
 install-udeb: install-deb
-	dh_testdir
-	dh_testroot
-	dh_clean -k
-	dh_installdirs
+	#dh_testdir
+	#dh_testroot
+	#dh_clean -k
+	#dh_installdirs
 
+	mkdir -p debian/udev-udeb/sbin
+	mkdir -p debian/udev-udeb/usr/bin
 	# Copy binaries that we want
 	cp -a udeb/udevd debian/udev-udeb/sbin
 	cp -a udeb/udevplug debian/udev-udeb/sbin
 	cp -a udeb/udevinfo debian/udev-udeb/usr/bin
 	cp -a udeb/udevmonitor debian/udev-udeb/usr/sbin
 
+	mkdir -p debian/udev-udeb/lib/udev
 	# Copy helpers that we want
 	cp -a udeb/extras/usb_device_name/usb_device_name debian/udev-udeb/lib/udev
 	cp -a udeb/extras/dvb_device_name/dvb_device_name debian/udev-udeb/lib/udev
@@ -170,6 +173,7 @@
 	# Copy all of the default devices from the deb
 	cp -a debian/udev/lib/udev/devices debian/udev-udeb/lib/udev
 
+	mkdir -p debian/udev-udeb/etc/udev
 	# Copy across the configuration file from the deb
 	cp -a debian/udev/etc/udev/udev.conf debian/udev-udeb/etc/udev
 
@@ -192,6 +196,7 @@
 	install -m 644 -o root -g root debian/rules.d/90-modprobe.rules \
 		debian/udev-udeb/etc/udev/rules.d
 
+	mkdir -p debian/udev-udeb/lib/debian-installer-startup.d/S02udev
 	# Startup script
 	install -m 755 -o root -g root debian/udev.installer-startup \
 		debian/udev-udeb/lib/debian-installer-startup.d/S02udev
@@ -205,33 +210,33 @@
 # Build architecture-dependent files here.
 binary-arch: DH_OPTIONS=-a
 binary-arch: build install-deb install-udeb
-	dh_testdir
-	dh_testroot
-	dh_installchangelogs ChangeLog
-	dh_installdocs
-	dh_installexamples
-	dh_installinit --no-start -- start 10 S .
-	dh_link
-	dh_strip
-	dh_compress
-	dh_fixperms -X/lib/udev/devices/
-	dh_installdeb
-	dh_shlibdeps
-	dh_gencontrol
-	dh_md5sums
-	dh_builddeb
+	#dh_testdir
+	#dh_testroot
+	#dh_installchangelogs ChangeLog
+	#dh_installdocs
+	#dh_installexamples
+	#dh_installinit --no-start -- start 10 S .
+	#dh_link
+	#dh_strip
+	#dh_compress
+	#dh_fixperms -X/lib/udev/devices/
+	#dh_installdeb
+	#dh_shlibdeps
+	#dh_gencontrol
+	#dh_md5sums
+	#dh_builddeb
 
 
 # Clean up the mess we made
 clean: clean-source unpatch
 clean-source:
-	dh_testdir
+	#dh_testdir
 	rm -f build-stamp
 
 	rm -rf udeb
 	-$(MAKE) clean V=true EXTRAS="$(EXTRAS)"
 
-	dh_clean 
+	#dh_clean 
 
 
 .PHONY: patch unpatch build install-deb install-udeb \
