diff -urN udev-0.060.orig.orig/extra/lndir.sh udev-0.060.orig/extra/lndir.sh
--- udev-0.060.orig.orig/extra/lndir.sh	2005-12-28 00:57:05.000000000 -0600
+++ udev-0.060.orig/extra/lndir.sh	2005-12-28 00:57:37.000000000 -0600
@@ -28,7 +28,7 @@
 
 # First, Canonicalise the pathnames
 
-DIRFROM="$(realpath $1)"
+DIRFROM="$(readlink -f $1)"
 
 if [ $# -ne 2 ];
 then
@@ -43,7 +43,7 @@
 
 echo "$DIRTO:"
 mkdir -p "$DIRTO"
-DIRTO="$(realpath $DIRTO)"
+DIRTO="$(readlink -f $DIRTO)"
 
 if [ ! -d "$DIRFROM" ]
 then
