--- udpcast-20120424/console.c.orig	2015-01-21 16:01:01.520139527 +0100
+++ udpcast-20120424/console.c	2015-01-21 16:02:03.285813096 +0100
@@ -3,6 +3,7 @@
 #include <stdio.h>
 #include <errno.h>
 #include <string.h>
+#include <sys/select.h>
 
 #include "console.h"
 #include "util.h"
@@ -11,7 +12,6 @@
 #ifndef __MINGW32__
 
 #include <termios.h>
-#include <sys/select.h>
 
 struct console_t {
     int fd; /* Filedescriptor for console, or -1 if disabled */
