--- udpcast/socklib.c.orig	Thu May  5 22:22:07 2005
+++ udpcast/socklib.c	Thu May  5 22:22:29 2005
@@ -175,13 +175,13 @@
 			   struct sockaddr *addr) {
     struct ifreq ifc;
     
+    struct sockaddr_in *iaddr = (struct sockaddr_in *) addr;
+    
     if(ifname)
 	strcpy(ifc.ifr_name, ifname);
     else if(addr_type != ADDR_TYPE_MCAST) {
 	udpc_fatal(1, "initSockAddr without ifname\n");
     }
-    
-    struct sockaddr_in *iaddr = (struct sockaddr_in *) addr;
     
     bzero ((char *) iaddr, sizeof(struct sockaddr_in));
     iaddr->sin_family = AF_INET;
