diff -u pci.lst~ pci.lst
--- foo/pci.lst~	Wed Aug 21 13:07:44 2002
+++ foo/pci.lst	Thu Mar 27 14:55:29 2003
@@ -930,7 +930,7 @@
 	10771020	unknown	qlogicisp	ISP1020
 	10771022	unknown	qlogicisp	ISP1022
 	10771080	unknown	qlogicisp	QLA1080
-	10771216	unknown	qlogicisp	QLA12160
+	10771216	unknown	qla1280		QLA12160
 	10771240	unknown	qlogicisp	QLA1240
 	10771280	unknown	qlogicisp	QLA1280
 	10772020	unknown	qlogicisp	ISP2020A
