--- discover1-data-1.2006.01.14/pci.lst.orig	2006-04-13 13:06:27.750900321 -0700
+++ discover1-data-1.2006.01.14/pci.lst	2006-04-13 13:09:39.101403228 -0700
@@ -149,7 +149,7 @@
 	10000013	scsi	sym53c8xx_2	53c875a
 	10000020	scsi	sym53c8xx_2	53c1010 Ultra3 SCSI Adapter
 	10000021	scsi	sym53c8xx_2	53c1010 66MHz  Ultra3 SCSI Adapter
-	10000030	scsi	mptscsih	53c1030 PCI-X Fusion-MPT Dual Ultra320 SCSI
+	10000030	scsi	mptspi		53c1030 PCI-X Fusion-MPT Dual Ultra320 SCSI
 	10000031	scsi	unknown	53c1030ZC PCI-X Fusion-MPT Dual Ultra320 SCSI
 	10000032	scsi	mptscsih	53c1035 PCI-X Fusion-MPT Dual Ultra320 SCSI
 	10000033	scsi	megaraid	1030ZC_53c1035 PCI-X Fusion-MPT Dual Ultra320 SCSI
