--- foo/pci.lst.orig	2003-04-21 23:27:23.000000000 -0600
+++ foo/pci.lst	2003-04-21 23:27:54.000000000 -0600
@@ -511,7 +511,7 @@
 	10280002	unknown	unknown	PowerEdge Expandable RAID Controller 3/Di
 	10280003	unknown	unknown	PowerEdge Expandable RAID Controller 3/Si
 	10280004	unknown	unknown	PowerEdge Expandable RAID Controller 3/Si
-	1028000a	unknown	unknown	PowerEdge Expandable RAID Controller 3/Si
+	1028000a	scsi	aacraid	PowerEdge Expandable RAID Controller 3/Si
 1029 Siemens Nixdorf IS
 102a LSI Logic
 	102a0000	unknown	unknown	HYDRA
