diff -urN discover-data-1.2002.08.21.orig/pci.lst discover-data-1.2002.08.21/pci.lst
--- discover-data-1.2002.08.21.orig/pci.lst	2002-08-21 13:07:44.000000000 -0600
+++ discover-data-1.2002.08.21/pci.lst	2003-05-31 16:06:15.000000000 -0600
@@ -686,7 +686,7 @@
 	10441012	unknown	unknown	Domino RAID Engine
 	1044a400	scsi	eata	SmartCache/Raid I-IV Controller
 	1044a500	bridge	ignore	PCI Bridge
-	1044a501	unknown	unknown	SmartRAID V Controller
+	1044a501	scsi	dpt_i2o	SmartRAID V Controller
 1045 OPTi Inc.
 	1045a0f8	usb	usb-ohci	82C750 [Vendetta] USB Controller 	1045c101	unknown	unknown	92C264
