--- discover-data-2.2006.10.29/pci.lst.orig	2006-10-26 01:06:27.000000000 -0700
+++ discover-data-2.2006.10.29/pci.lst	2007-02-09 19:52:32.000000000 -0800
@@ -188,13 +188,13 @@
 	10000013	scsi	sym53c8xx_2	53c875a
 	10000020	scsi	sym53c8xx_2	53c1010 Ultra3 SCSI Adapter
 	10000021	scsi	sym53c8xx_2	53c1010 66MHz  Ultra3 SCSI Adapter
-	10000030	scsi	mptscsih	53c1030 PCI-X Fusion-MPT Dual Ultra320 SCSI
+	10000030	scsi	mptspi		53c1030 PCI-X Fusion-MPT Dual Ultra320 SCSI
 	10000031	scsi	unknown	53c1030ZC PCI-X Fusion-MPT Dual Ultra320 SCSI
 	10000032	scsi	mptscsih	53c1035 PCI-X Fusion-MPT Dual Ultra320 SCSI
 	10000033	scsi	megaraid	1030ZC_53c1035 PCI-X Fusion-MPT Dual Ultra320 SCSI
 	10000040	scsi	mptscsih	53c1035 PCI-X Fusion-MPT Dual Ultra320 SCSI
 	10000041	scsi	unknown	53C1035ZC PCI-X Fusion-MPT Dual Ultra320 SCSI
-	10000050	scsi	mptscsih	SAS1064 PCI-X Fusion-MPT SAS
+	10000050	scsi	mptsas		SAS1064 PCI-X Fusion-MPT SAS
 	10000054	scsi	mptscsih	SAS1068 PCI-X Fusion-MPT SAS
 	10000056	unknown	unknown	SAS1064E PCI-Express Fusion-MPT SAS
 	10000058	unknown	unknown	SAS1068E PCI-Express Fusion-MPT SAS
@@ -2529,8 +2529,8 @@
 	10de0057	ethernet	forcedeth	CK804 Ethernet Controller
 	10de0058	unknown	unknown	CK804 AC'97 Modem
 	10de0059	sound	unknown	CK804 AC'97 Audio Controller
-	10de005a	usb	unknown	CK804 USB Controller
-	10de005b	usb	unknown	CK804 USB Controller
+	10de005a	usb	ohci_hcd	CK804 USB Controller
+	10de005b	usb	ehci_hcd	CK804 USB Controller
 	10de005c	unknown	unknown	CK804 PCI Bridge
 	10de005d	unknown	unknown	CK804 PCIE Bridge
 	10de005e	unknown	unknown	CK804 Memory Controller
@@ -7151,7 +7151,7 @@
 	80861093	unknown	unknown	PRO/100 VM Network Connection
 	80861094	unknown	unknown	PRO/100 VE Network Connection
 	80861095	unknown	unknown	PRO/100 VE Network Connection
-	80861096	unknown	unknown	80003ES2LAN Gigabit Ethernet Controller (Copper)
+	80861096	ethernet	e1000	80003ES2LAN Gigabit Ethernet Controller (Copper)
 	80861097	unknown	unknown	631xESB/632xESB DPT LAN Controller (Fiber)
 	80861098	unknown	unknown	80003ES2LAN Gigabit Ethernet Controller (Serdes)
 	80861099	ethernet	unknown	82546GB Gigabit Ethernet Controller (Copper)
@@ -7436,11 +7436,11 @@
 	80862681	unknown	unknown	631xESB/632xESB SATA Storage Controller AHCI
 	80862682	unknown	unknown	631xESB/632xESB SATA Storage Controller RAID
 	80862683	unknown	unknown	631xESB/632xESB SATA Storage Controller RAID
-	80862688	unknown	unknown	631xESB/632xESB/3100 Chipset UHCI USB Controller #1
-	80862689	unknown	unknown	631xESB/632xESB/3100 Chipset UHCI USB Controller #2
-	8086268a	unknown	unknown	631xESB/632xESB/3100 Chipset UHCI USB Controller #3
-	8086268b	unknown	unknown	631xESB/632xESB/3100 Chipset UHCI USB Controller #4
-	8086268c	unknown	unknown	631xESB/632xESB/3100 Chipset EHCI USB2 Controller
+	80862688	usb	uhci_hcd	631xESB/632xESB/3100 Chipset UHCI USB Controller #1
+	80862689	usb	uhci_hcd	631xESB/632xESB/3100 Chipset UHCI USB Controller #2
+	8086268a	usb	uhci_hcd	631xESB/632xESB/3100 Chipset UHCI USB Controller #3
+	8086268b	usb	uhci_hcd	631xESB/632xESB/3100 Chipset UHCI USB Controller #4
+	8086268c	usb	ehci_hcd	631xESB/632xESB/3100 Chipset EHCI USB2 Controller
 	80862690	unknown	unknown	631xESB/632xESB/3100 Chipset PCI Express Root Port 1
 	80862692	unknown	unknown	631xESB/632xESB/3100 Chipset PCI Express Root Port 2
 	80862694	unknown	unknown	631xESB/632xESB/3100 Chipset PCI Express Root Port 3
