diff -urN discover-data-1.2002.08.21.orig/pci.lst discover-data-1.2002.08.21/pci.lst
--- discover-data-1.2002.08.21.orig/pci.lst	2003-06-15 11:15:33.000000000 -0600
+++ discover-data-1.2002.08.21/pci.lst	2003-06-15 11:15:50.000000000 -0600
@@ -66,7 +66,13 @@
 	10000012	scsi	sym53c8xx	53c895a
 	10000020	scsi	sym53c8xx	53c1010 Ultra3 SCSI Adapter
 	10000021	scsi	sym53c8xx	53c1010 66MHz Ultra3 SCSI Adapter
+	10000030	scsi	mptscsih	LSI Logic|53c1030
 	1000008f	scsi	ncr53c8xx	53c875J
+	10000621	scsi	mptscsih	LSI Logic|LSIFC909
+	10000622	scsi	mptscsih	LSI Logic|LSIFC929
+	10000624	scsi	mptscsih	LSI Logic|LSIFC919
+	10000626	scsi	mptscsih	LSI Logic|LSIFC929X
+	10000628	scsi	mptscsih	LSI Logic|LSIFC919X
 	10000701	scsi	yellowfin	83C885
 	10000702	scsi	yellowfin	Yellowfin G-NIC gigabit ethernet
 	10000901	unknown	unknown	61C102
