--- discover-1.5/discover/discover.c.orig	2003-06-01 17:17:55.000000000 -0600
+++ discover-1.5/discover/discover.c	2003-06-01 17:24:48.000000000 -0600
@@ -564,26 +564,26 @@
 static void
 usage(int status)
 {
-    printf("Usage: %s [OPTIONS] DEVICE [DEVICE...]\n\
-\n\
-      --enable=BUS[,BUS...]     Enable probing of BUS.\n\
-      --enable-all              Enable probing of all buses.\n\
-      --disable=BUS[,BUS...]    Disable probing of BUS.\n\
-      --disable-all             Disable probing of all buses.\n\
-  -f, --format=STRING           Format output according to STRING. STRING
-                                is a printf-like string, where %%V expands
-                                to the vendor, %%M expands to the model,
-                                %%m expands to the module to be loaded, %%d
-                                expands to the path of the device file,
-                                %%S expands to the name of the X server
-                                to be used, and %%D expands to the name
-                                of X server driver to be used.\n\
-      --vendor                  Print vendor.\n\
-      --model                   Print model.\n\
-      --module                  Print module to be loaded.\n\
-      --device                  Print path of device file.\n\
-      --xserver                 Print X server.\n\
-      --xdriver                 Print X server driver.\n", program_name);
+    printf("Usage: %s [OPTIONS] DEVICE [DEVICE...]\n", program_name);
+    printf("\n");
+    printf("      --enable=BUS[,BUS...]     Enable probing of BUS.\n");
+    printf("      --enable-all              Enable probing of all buses.\n");
+    printf("      --disable=BUS[,BUS...]    Disable probing of BUS.\n");
+    printf("      --disable-all             Disable probing of all buses.\n");
+    printf("  -f, --format=STRING           Format output according to STRING. STRING\n");
+    printf("                                is a printf-like string, where %%V expands\n");
+    printf("                                to the vendor, %%M expands to the model,\n");
+    printf("                                %%m expands to the module to be loaded, %%d\n");
+    printf("                                expands to the path of the device file,\n");
+    printf("                                %%S expands to the name of the X server\n");
+    printf("                                to be used, and %%D expands to the name\n");
+    printf("                                of X server driver to be used.\n");
+    printf("      --vendor                  Print vendor.\n");
+    printf("      --model                   Print model.\n");
+    printf("      --module                  Print module to be loaded.\n");
+    printf("      --device                  Print path of device file.\n");
+    printf("      --xserver                 Print X server.\n");
+    printf("      --xdriver                 Print X server driver.\n");
     exit(0);
 }
 
