--- dosfsck/dosfsck.h.moo	Wed Sep  3 10:38:29 2003
+++ dosfsck/dosfsck.h	Wed Sep  3 10:38:42 2003
@@ -14,14 +14,22 @@
 #define _LINUX_STRING_H_	/* hack to avoid inclusion of <linux/string.h>*/
 #include <linux/msdos_fs.h>
 
-/* 2.1 kernels use le16_to_cpu() type functions for CF_LE_W & Co., but don't
- * export this macros, only __le16_to_cpu(). */
-#ifndef le16_to_cpu
-#define le16_to_cpu	__le16_to_cpu
-#define le32_to_cpu	__le32_to_cpu
-#define cpu_to_le16	__cpu_to_le16
-#define cpu_to_le32	__cpu_to_le32
-#endif
+#if __BYTE_ORDER == __BIG_ENDIAN
+#include <byteswap.h>
+
+#define CF_LE_W(v) bswap_16(v)
+#define CF_LE_L(v) bswap_32(v)
+#define CT_LE_W(v) CF_LE_W(v)
+#define CT_LE_L(v) CF_LE_L(v)
+    
+#else
+
+#define CF_LE_W(v) (v)
+#define CF_LE_L(v) (v)
+#define CT_LE_W(v) (v)
+#define CT_LE_L(v) (v)
+
+#endif /* __BIG_ENDIAN */
 
 #define VFAT_LN_ATTR (ATTR_RO | ATTR_HIDDEN | ATTR_SYS | ATTR_VOLUME)
 
