--- mkdosfs/mkdosfs.c.loopsize	2003-02-19 00:50:30.000000000 -0500
+++ mkdosfs/mkdosfs.c	2003-02-19 00:50:38.000000000 -0500
@@ -675,7 +675,6 @@
       /* Can we get the loop geometry? This is in 512 byte blocks, always? */
       if (ioctl (dev, BLKGETSIZE, &loop_size)) 
 	die ("unable to get loop geometry for '%s'");
-      loop_size = loop_size >> 1;
 
       switch (loop_size)  /* Assuming the loop device -> floppy later */
 	{
