--- jfsutils-1.1.3/libfs/devices.c.orig	2003-11-30 11:53:29.000000000 -0700
+++ jfsutils-1.1.3/libfs/devices.c	2003-11-30 11:55:08.000000000 -0700
@@ -31,7 +31,7 @@
 #include <sys/ioctl.h>
 
 #if defined(__linux__) && defined(_IO) && !defined(BLKGETSIZE64)
-#define BLKGETSIZE64 _IOR(0x12, 114, sizeof(uint64_t))
+#define BLKGETSIZE64 _IOR(0x12,114,size_t)
 #endif
 #if defined(__linux__) && defined(_IO) && !defined(BLKGETSIZE)
 #define BLKGETSIZE _IO(0x12,96)	/* return device size (sectors) */
