diff -Naur linux-2.4.7/Documentation/Changes jfs-2.4.7-1.0.11/Documentation/Changes
--- linux-2.4.7/Documentation/Changes	Thu Dec  6 14:13:50 2001
+++ jfs-2.4.7-1.0.11/Documentation/Changes	Mon Dec 17 16:25:53 2001
@@ -54,6 +54,7 @@
 o  util-linux             2.10o                   # fdformat --version
 o  modutils               2.4.2                   # insmod -V
 o  e2fsprogs              1.19                    # tune2fs
+o  jfsprogs               1.0.10                  # fsck.jfs -V
 o  reiserfsprogs          3.x.0j                  # reiserfsck 2>&1|grep reiserfsprogs
 o  pcmcia-cs              3.1.21                  # cardmgr -V
 o  PPP                    2.4.0                   # pppd --version
@@ -165,6 +166,16 @@
 The latest version of e2fsprogs fixes several bugs in fsck and
 debugfs.  Obviously, it's a good idea to upgrade.
 
+JFSprogs
+--------
+
+The jfsprogs package contains the utilities for the file system.
+The following utilities are available:
+o fsck.jfs - initiate replay of the transaction log, and check
+  and repair a JFS formatted partition.
+o mkfs.jfs - create a JFS formatted partition.
+o other file system utilities are also available in this package.
+
 Reiserfsprogs
 -------------
 
@@ -313,6 +324,10 @@
 ---------
 o  <http://prdownloads.sourceforge.net/e2fsprogs/e2fsprogs-1.19.tar.gz>
 o  <http://prdownloads.sourceforge.net/e2fsprogs/e2fsprogs-1.19-0.src.rpm>
+
+JFSprogs
+--------
+o  <http://oss.software.ibm.com/jfs>
 
 Reiserfsprogs
 -------------
diff -Naur linux-2.4.7/Documentation/Configure.help jfs-2.4.7-1.0.11/Documentation/Configure.help
--- linux-2.4.7/Documentation/Configure.help	Wed Aug  1 09:17:54 2001
+++ jfs-2.4.7-1.0.11/Documentation/Configure.help	Mon Dec 17 16:25:53 2001
@@ -12551,6 +12551,20 @@
   hard drives and ADFS-formatted floppy disks. This is experimental
   codes, so if you're unsure, say N.
 
+JFS filesystem support
+CONFIG_JFS_FS
+  This is a port of IBM's Journaled Filesystem .  More information is
+  available in the file Documentation/filesystems/jfs.txt.
+
+  If you do not intend to use the JFS filesystem, say N.
+
+JFS Debugging
+CONFIG_JFS_DEBUG
+  If you are experiencing any problems with the JFS filesystem, say
+  Y here.  This will result in additional debugging messages to be
+  written to the system log.  Under normal circumstances, this
+  results in very little overhead.
+
 /dev/pts file system for Unix98 PTYs
 CONFIG_DEVPTS_FS
   You should say Y here if you said Y to "Unix98 PTY support" above.
diff -Naur linux-2.4.7/Documentation/filesystems/00-INDEX jfs-2.4.7-1.0.11/Documentation/filesystems/00-INDEX
--- linux-2.4.7/Documentation/filesystems/00-INDEX	Wed Aug  1 09:17:54 2001
+++ jfs-2.4.7-1.0.11/Documentation/filesystems/00-INDEX	Mon Dec 17 16:25:53 2001
@@ -22,6 +22,8 @@
 	- info and mount options for the OS/2 HPFS.
 isofs.txt
 	- info and mount options for the ISO 9660 (CDROM) filesystem.
+jfs.txt
+	- info and mount options for the JFS filesystem.
 ncpfs.txt
 	- info on Novell Netware(tm) filesystem using NCP protocol.
 ntfs.txt
diff -Naur linux-2.4.7/MAINTAINERS jfs-2.4.7-1.0.11/MAINTAINERS
--- linux-2.4.7/MAINTAINERS	Wed Aug  1 09:17:54 2001
+++ jfs-2.4.7-1.0.11/MAINTAINERS	Mon Dec 17 16:25:53 2001
@@ -741,6 +741,13 @@
 W:	http://www.developer.axis.com/software/jffs/
 S:	Maintained
 
+JFS FILESYSTEM
+P:	Dave Kleikamp
+M:	shaggy@austin.ibm.com
+L:	jfs-discussion@oss.software.ibm.com
+W:	http://oss.software.ibm.com/developerworks/opensource/jfs/
+S:	Supported
+
 JOYSTICK DRIVER
 P:	Vojtech Pavlik
 M:	vojtech@suse.cz
diff -Naur linux-2.4.7/fs/Config.in jfs-2.4.7-1.0.11/fs/Config.in
--- linux-2.4.7/fs/Config.in	Wed Aug  1 09:17:55 2001
+++ jfs-2.4.7-1.0.11/fs/Config.in	Mon Dec 17 16:25:54 2001
@@ -37,6 +37,9 @@
 tristate 'ISO 9660 CDROM file system support' CONFIG_ISO9660_FS
 dep_mbool '  Microsoft Joliet CDROM extensions' CONFIG_JOLIET $CONFIG_ISO9660_FS
 
+tristate 'JFS filesystem support' CONFIG_JFS_FS
+dep_mbool '  JFS debugging' CONFIG_JFS_DEBUG $CONFIG_JFS_FS
+
 tristate 'Minix fs support' CONFIG_MINIX_FS
 
 tristate 'FreeVxFS file system support (VERITAS VxFS(TM) compatible)' CONFIG_VXFS_FS
diff -Naur linux-2.4.7/fs/Makefile jfs-2.4.7-1.0.11/fs/Makefile
--- linux-2.4.7/fs/Makefile	Wed Aug  1 09:17:55 2001
+++ jfs-2.4.7-1.0.11/fs/Makefile	Mon Dec 17 16:25:54 2001
@@ -62,6 +62,7 @@
 subdir-$(CONFIG_REISERFS_FS)	+= reiserfs
 subdir-$(CONFIG_DEVPTS_FS)	+= devpts
 subdir-$(CONFIG_SUN_OPENPROMFS)	+= openpromfs
+subdir-$(CONFIG_JFS_FS)		+= jfs
 
 
 obj-$(CONFIG_BINFMT_AOUT)	+= binfmt_aout.o
diff -Naur linux-2.4.7/fs/nls/Config.in jfs-2.4.7-1.0.11/fs/nls/Config.in
--- linux-2.4.7/fs/nls/Config.in	Wed Aug  1 09:17:55 2001
+++ jfs-2.4.7-1.0.11/fs/nls/Config.in	Mon Dec 17 16:25:54 2001
@@ -12,7 +12,7 @@
 # msdos and Joliet want NLS
 if [ "$CONFIG_JOLIET" = "y" -o "$CONFIG_FAT_FS" != "n" \
 	-o "$CONFIG_NTFS_FS" != "n" -o "$CONFIG_NCPFS_NLS" = "y" \
-	-o "$CONFIG_SMB_NLS" = "y" ]; then
+	-o "$CONFIG_SMB_NLS" = "y" -o "$CONFIG_JFS_FS" != "n" ]; then
   define_bool CONFIG_NLS y
 else
   define_bool CONFIG_NLS n
