--- linux-2.4.20.orig/Makefile	2003-01-12 14:26:47.000000000 -0700
+++ linux-2.4.20/Makefile	2003-01-12 14:34:56.000000000 -0700
@@ -92,7 +92,7 @@
 CPPFLAGS := -D__KERNEL__ -I$(HPATH)
 CPPFLAGS += $(patsubst %,-I%,$(CROSS_COMPILE_INC))
 
-CFLAGS := $(CPPFLAGS) -Wall -Wstrict-prototypes -Wno-trigraphs -O2 \
+CFLAGS := $(CPPFLAGS) -Wall -Wstrict-prototypes -Wno-trigraphs -Os \
 	  -fno-strict-aliasing -fno-common
 CFLAGS += -g
 ifndef CONFIG_FRAME_POINTER
