diff -urN linux.orig/drivers/net/bcm/b57um.c linux/drivers/net/bcm/b57um.c
--- linux.orig/drivers/net/bcm/b57um.c	2002-11-06 22:35:52.000000000 -0700
+++ linux/drivers/net/bcm/b57um.c	2002-11-06 22:42:14.000000000 -0700
@@ -868,7 +868,7 @@
 }
 
 
-static void __devexit
+static void 
 bcm5700_remove_one (struct pci_dev *pdev)
 {
 	struct net_device *dev = pci_get_drvdata (pdev);
@@ -2231,7 +2231,7 @@
 }
 
 
-static void __exit bcm5700_cleanup_module (void)
+static void bcm5700_cleanup_module (void)
 {
 	pci_unregister_driver(&bcm5700_pci_driver);
 }
