--- linux-2.6.20.4/drivers/net/depca.c.orig	2007-04-04 12:39:00.000000000 +0200
+++ linux-2.6.20.4/drivers/net/depca.c	2007-04-04 12:39:17.000000000 +0200
@@ -1491,8 +1491,9 @@
 		depca_io_ports[i].device = pldev;
 
 		if (platform_device_add(pldev)) {
-			platform_device_put(pldev);
 			depca_io_ports[i].device = NULL;
+			pldev->dev.platform_data = NULL;
+			platform_device_put(pldev);
 			continue;
 		}
 
