--- linux-2.4.20.orig/arch/ia64/kernel/smpboot.c	2002-11-28 16:53:09.000000000 -0700
+++ linux-2.4.20/arch/ia64/kernel/smpboot.c	2003-01-23 15:23:42.000000000 -0700
@@ -89,7 +89,7 @@
 struct smp_boot_data smp_boot_data __initdata;
 
 /* Set when the idlers are all forked */
-volatile int smp_threads_ready;
+int smp_threads_ready;
 
 unsigned long ap_wakeup_vector = -1; /* External Int use to wakeup APs */
 
