--- linux-2.4.20/net/core/rtnetlink.c.orig	2003-06-08 19:13:06.000000000 -0600
+++ linux-2.4.20/net/core/rtnetlink.c	2003-06-08 19:13:32.000000000 -0600
@@ -394,7 +394,7 @@
  * Malformed skbs with wrong lengths of messages are discarded silently.
  */
 
-extern __inline__ int rtnetlink_rcv_skb(struct sk_buff *skb)
+__inline__ int rtnetlink_rcv_skb(struct sk_buff *skb)
 {
 	int err;
 	struct nlmsghdr * nlh;
