--- linux-2.6.24.4/Makefile.orig	2008-04-08 16:24:22.000000000 +0200
+++ linux-2.6.24.4/Makefile	2008-04-08 17:21:02.000000000 +0200
@@ -333,6 +333,11 @@ KBUILD_CPPFLAGS := -D__KERNEL__ $(LINUXI
 KBUILD_CFLAGS   := -Wall -Wundef -Wstrict-prototypes -Wno-trigraphs \
 		   -fno-strict-aliasing -fno-common \
 		   -Werror-implicit-function-declaration
+
+ifeq ($(shell ($(CC) -fno-tree-scev-cprop -shared -o /tmp/check_cc -xc /dev/null >/dev/null 2>&1 && echo ok && rm -f /tmp/check_cc)), ok)
+KBUILD_CFLAGS	+= -fno-tree-scev-cprop
+endif
+
 KBUILD_AFLAGS   := -D__ASSEMBLY__
 
 # Read KERNELRELEASE from include/config/kernel.release (if it exists)
