diff -urN linux-clean/drivers/char/pc_keyb.c pc_keyb/drivers/char/pc_keyb.c
--- linux-clean/drivers/char/pc_keyb.c	Tue Aug 14 16:49:50 2001
+++ pc_keyb/drivers/char/pc_keyb.c	Wed Sep  5 10:01:45 2001
@@ -723,6 +723,13 @@
 {
 	int status;
 
+#ifdef __ia64__
+	if (kbd_read_status() == 0xff && kbd_read_input() == 0xff) {
+		kbd_exists = 0;
+		return "No keyboard controller preset";
+	}
+#endif
+
 	/*
 	 * Test the keyboard interface.
 	 * This seems to be the only way to get it going.
@@ -829,6 +836,10 @@
 		char *msg = initialize_kbd();
 		if (msg)
 			printk(KERN_WARNING "initialize_kbd: %s\n", msg);
+#ifdef __ia64__
+		if (!kbd_exists)
+			return;
+#endif
 	}
 
 #if defined CONFIG_PSMOUSE
