diff -urN linux.orig/arch/i386/vmlinux.lds linux/arch/i386/vmlinux.lds
--- linux.orig/arch/i386/vmlinux.lds	Mon Jul  2 15:40:14 2001
+++ linux/arch/i386/vmlinux.lds	Wed Jan  2 23:59:36 2002
@@ -65,13 +65,6 @@
 	}
   _end = . ;
 
-  /* Sections to be discarded */
-  /DISCARD/ : {
-	*(.text.exit)
-	*(.data.exit)
-	*(.exitcall.exit)
-	}
-
   /* Stabs debugging sections.  */
   .stab 0 : { *(.stab) }
   .stabstr 0 : { *(.stabstr) }
