diff -urN linux-2.4.20.orig/drivers/block/rd.c linux-2.4.20/drivers/block/rd.c
--- linux-2.4.20.orig/drivers/block/rd.c	2002-11-28 16:53:12.000000000 -0700
+++ linux-2.4.20/drivers/block/rd.c	2003-01-13 22:21:57.000000000 -0700
@@ -97,7 +97,7 @@
  * behaviour. The default is still BLOCK_SIZE (needed by rd_load_image that
  * supposes the filesystem in the image uses a BLOCK_SIZE blocksize).
  */
-int rd_blocksize = BLOCK_SIZE;			/* blocksize of the RAM disks */
+int rd_blocksize = 4096;			/* blocksize of the RAM disks */
 
 /*
  * Copyright (C) 2000 Linus Torvalds.
@@ -478,14 +478,8 @@
 {
 	return ramdisk_size(str);
 }
-static int __init ramdisk_blocksize(char *str)
-{
-	rd_blocksize = simple_strtol(str,NULL,0);
-	return 1;
-}
 __setup("ramdisk=", ramdisk_size);
 __setup("ramdisk_size=", ramdisk_size2);
-__setup("ramdisk_blocksize=", ramdisk_blocksize);
 #endif
 
 /* options - modular */
