diff -Naur linux-2.4.18/Documentation/Changes linux-2.4.18-jfs/Documentation/Changes
--- linux-2.4.18/Documentation/Changes	Mon Mar 11 13:56:57 2002
+++ linux-2.4.18-jfs/Documentation/Changes	Mon Mar 11 13:58:11 2002
@@ -54,6 +54,7 @@
 o  util-linux             2.10o                   # fdformat --version
 o  modutils               2.4.2                   # insmod -V
 o  e2fsprogs              1.25                    # tune2fs
+o  jfsutils               1.0.14                  # fsck.jfs -V
 o  reiserfsprogs          3.x.0j                  # reiserfsck 2>&1|grep reiserfsprogs
 o  pcmcia-cs              3.1.21                  # cardmgr -V
 o  PPP                    2.4.0                   # pppd --version
@@ -171,6 +172,16 @@
 The latest version of e2fsprogs fixes several bugs in fsck and
 debugfs.  Obviously, it's a good idea to upgrade.
 
+JFSutils
+--------
+
+The jfsutils package contains the utilities for the file system.
+The following utilities are available:
+o fsck.jfs - initiate replay of the transaction log, and check
+  and repair a JFS formatted partition.
+o mkfs.jfs - create a JFS formatted partition.
+o other file system utilities are also available in this package.
+
 Reiserfsprogs
 -------------
 
@@ -318,6 +329,10 @@
 E2fsprogs
 ---------
 o  <http://prdownloads.sourceforge.net/e2fsprogs/e2fsprogs-1.25.tar.gz>
+
+JFSutils
+---------
+o  <http://oss.software.ibm.com/jfs>
 
 Reiserfsprogs
 -------------
diff -Naur linux-2.4.18/Documentation/Configure.help linux-2.4.18-jfs/Documentation/Configure.help
--- linux-2.4.18/Documentation/Configure.help	Mon Mar 11 13:56:58 2002
+++ linux-2.4.18-jfs/Documentation/Configure.help	Mon Mar 11 13:58:11 2002
@@ -15012,6 +15012,25 @@
   hard drives and ADFS-formatted floppy disks. This is experimental
   codes, so if you're unsure, say N.
 
+JFS filesystem support
+CONFIG_JFS_FS
+  This is a port of IBM's Journaled Filesystem .  More information is
+  available in the file Documentation/filesystems/jfs.txt.
+
+  If you do not intend to use the JFS filesystem, say N.
+
+JFS Debugging
+CONFIG_JFS_DEBUG
+  If you are experiencing any problems with the JFS filesystem, say
+  Y here.  This will result in additional debugging messages to be
+  written to the system log.  Under normal circumstances, this
+  results in very little overhead.
+
+JFS Statistics
+CONFIG_JFS_STATISTICS
+  Enabling this option will cause statistics from the JFS file system
+  to be made available to the user in the /proc/fs/jfs/ directory.
+
 /dev/pts file system for Unix98 PTYs
 CONFIG_DEVPTS_FS
   You should say Y here if you said Y to "Unix98 PTY support" above.
diff -Naur linux-2.4.18/Documentation/filesystems/00-INDEX linux-2.4.18-jfs/Documentation/filesystems/00-INDEX
--- linux-2.4.18/Documentation/filesystems/00-INDEX	Mon Mar 11 13:56:58 2002
+++ linux-2.4.18-jfs/Documentation/filesystems/00-INDEX	Mon Mar 11 13:58:11 2002
@@ -22,6 +22,8 @@
 	- info and mount options for the OS/2 HPFS.
 isofs.txt
 	- info and mount options for the ISO 9660 (CDROM) filesystem.
+jfs.txt
+	- info and mount options for the JFS filesystem.
 ncpfs.txt
 	- info on Novell Netware(tm) filesystem using NCP protocol.
 ntfs.txt
diff -Naur linux-2.4.18/MAINTAINERS linux-2.4.18-jfs/MAINTAINERS
--- linux-2.4.18/MAINTAINERS	Mon Mar 11 13:56:57 2002
+++ linux-2.4.18-jfs/MAINTAINERS	Mon Mar 11 13:58:11 2002
@@ -844,6 +844,13 @@
 W:	http://sources.redhat.com/jffs2/
 S:	Maintained
 
+JFS FILESYSTEM
+P:	Dave Kleikamp
+M:	shaggy@austin.ibm.com
+L:	jfs-discussion@oss.software.ibm.com
+W:	http://oss.software.ibm.com/developerworks/opensource/jfs/
+S:	Supported
+
 JOYSTICK DRIVER
 P:	Vojtech Pavlik
 M:	vojtech@suse.cz
diff -Naur linux-2.4.18/fs/Config.in linux-2.4.18-jfs/fs/Config.in
--- linux-2.4.18/fs/Config.in	Mon Mar 11 13:56:58 2002
+++ linux-2.4.18-jfs/fs/Config.in	Mon Mar 11 13:58:11 2002
@@ -51,6 +51,10 @@
 dep_mbool '  Microsoft Joliet CDROM extensions' CONFIG_JOLIET $CONFIG_ISO9660_FS
 dep_mbool '  Transparent decompression extension' CONFIG_ZISOFS $CONFIG_ISO9660_FS
 
+tristate 'JFS filesystem support' CONFIG_JFS_FS
+dep_mbool '  JFS debugging' CONFIG_JFS_DEBUG $CONFIG_JFS_FS
+dep_mbool '  JFS statistics' CONFIG_JFS_STATISTICS $CONFIG_JFS_FS
+
 tristate 'Minix fs support' CONFIG_MINIX_FS
 
 tristate 'FreeVxFS file system support (VERITAS VxFS(TM) compatible)' CONFIG_VXFS_FS
diff -Naur linux-2.4.18/fs/Makefile linux-2.4.18-jfs/fs/Makefile
--- linux-2.4.18/fs/Makefile	Mon Mar 11 13:56:58 2002
+++ linux-2.4.18-jfs/fs/Makefile	Mon Mar 11 13:58:11 2002
@@ -67,6 +67,7 @@
 subdir-$(CONFIG_REISERFS_FS)	+= reiserfs
 subdir-$(CONFIG_DEVPTS_FS)	+= devpts
 subdir-$(CONFIG_SUN_OPENPROMFS)	+= openpromfs
+subdir-$(CONFIG_JFS_FS)		+= jfs
 
 
 obj-$(CONFIG_BINFMT_AOUT)	+= binfmt_aout.o
diff -Naur linux-2.4.18/fs/nls/Config.in linux-2.4.18-jfs/fs/nls/Config.in
--- linux-2.4.18/fs/nls/Config.in	Mon Mar 11 13:56:58 2002
+++ linux-2.4.18-jfs/fs/nls/Config.in	Mon Mar 11 13:58:11 2002
@@ -12,7 +12,7 @@
 # msdos and Joliet want NLS
 if [ "$CONFIG_JOLIET" = "y" -o "$CONFIG_FAT_FS" != "n" \
 	-o "$CONFIG_NTFS_FS" != "n" -o "$CONFIG_NCPFS_NLS" = "y" \
-	-o "$CONFIG_SMB_NLS" = "y" ]; then
+	-o "$CONFIG_SMB_NLS" = "y" -o "$CONFIG_JFS_FS" != "n" ]; then
   define_bool CONFIG_NLS y
 else
   define_bool CONFIG_NLS n
