--- clean/arch/ia64/kernel/setup.c	Tue Jul 31 11:30:08 2001
+++ patched/arch/ia64/kernel/setup.c	Thu Nov 15 15:04:34 2001
@@ -344,7 +344,12 @@
 
 #ifdef CONFIG_VT
 # if defined(CONFIG_VGA_CONSOLE)
-	conswitchp = &vga_con;
+	if (!(efi_page_is_ram(0xa0000) == 1))
+		conswitchp = &vga_con;
+#  if defined(CONFIG_DUMMY_CONSOLE)
+	else
+		conswitchp = &dummy_con;
+#  endif
 # elif defined(CONFIG_DUMMY_CONSOLE)
 	conswitchp = &dummy_con;
 # endif
