diff -urN linux.orig/arch/ia64/vmlinux.lds.S linux/arch/ia64/vmlinux.lds.S
--- linux.orig/arch/ia64/vmlinux.lds.S	Thu Apr  4 15:32:35 2002
+++ linux/arch/ia64/vmlinux.lds.S	Thu Apr  4 15:33:20 2002
@@ -8,15 +8,6 @@
 ENTRY(phys_start)
 SECTIONS
 {
-  /* Sections to be discarded */
-  /DISCARD/ : {
-	*(.text.exit)
-	*(.data.exit)
-	*(*.text.exit)
-	*(*.data.exit)
-	*(.exitcall.exit)
-	}
-
   v = PAGE_OFFSET;	/* this symbol is here to make debugging easier... */
   phys_start = _start - PAGE_OFFSET;
 
