--- linux/Makefile.orig	2007-10-29 22:50:54.000000000 +0100
+++ linux/Makefile	2007-10-29 22:52:26.000000000 +0100
@@ -185,7 +185,7 @@ CONFIG_SHELL := $(shell if [ -x "$$BASH"
 
 HOSTCC  	= gcc
 HOSTCXX  	= g++
-HOSTCFLAGS	= -Wall -Wstrict-prototypes -O2 -fomit-frame-pointer
+HOSTCFLAGS	= -Wall -Wstrict-prototypes -O2 -fomit-frame-pointer $(call cc-option, -fno-stack-protector)
 HOSTCXXFLAGS	= -O2
 
 # 	Decide whether to build built-in, modular, or both.
@@ -321,11 +321,11 @@ PERL		= perl
 CHECK		= sparse
 
 CHECKFLAGS     := -D__linux__ -Dlinux -D__STDC__ -Dunix -D__unix__ -Wbitwise $(CF)
-MODFLAGS	= -DMODULE
+MODFLAGS	= -DMODULE $(call cc-option, -fno-stack-protector)
 CFLAGS_MODULE   = $(MODFLAGS)
 AFLAGS_MODULE   = $(MODFLAGS)
 LDFLAGS_MODULE  = -r
-CFLAGS_KERNEL	=
+CFLAGS_KERNEL	= $(call cc-option, -fno-stack-protector)
 AFLAGS_KERNEL	=
 
 
@@ -339,7 +339,7 @@ CPPFLAGS        := -D__KERNEL__ $(LINUXI
 
 CFLAGS 		:= -Wall -Wundef -Wstrict-prototypes -Wno-trigraphs \
 	  	   -fno-strict-aliasing -fno-common \
-		   -ffreestanding
+		   -ffreestanding $(call cc-option, -fno-stack-protector)
 AFLAGS		:= -D__ASSEMBLY__
 
 # Read KERNELRELEASE from .kernelrelease (if it exists)
