diff -urN linux.orig/Makefile linux/Makefile
--- linux.orig/Makefile	Wed Jan  2 23:25:59 2002
+++ linux/Makefile	Wed Jan  2 23:26:37 2002
@@ -106,7 +106,7 @@
 CPPFLAGS := -D__KERNEL__ -I$(HPATH)
 CPPFLAGS += $(patsubst %,-I%,$(CROSS_COMPILE_INC))
 
-CFLAGS := $(CPPFLAGS) -Wall -Wstrict-prototypes -Wno-trigraphs -O2 \
+CFLAGS := $(CPPFLAGS) -Wall -Wstrict-prototypes -Wno-trigraphs -Os \
 	  -fno-strict-aliasing -fno-common
 ifndef CONFIG_FRAME_POINTER
 CFLAGS += -fomit-frame-pointer
