diff -urN linux.orig/Documentation/Changes linux/Documentation/Changes
--- linux.orig/Documentation/Changes	2002-09-15 18:47:32.000000000 -0600
+++ linux/Documentation/Changes	2002-09-15 19:02:01.000000000 -0600
@@ -54,7 +54,6 @@
 o  util-linux             2.10o                   # fdformat --version
 o  modutils               2.4.2                   # insmod -V
 o  e2fsprogs              1.25                    # tune2fs
-o  jfsutils               1.0.14                  # fsck.jfs -V
 o  reiserfsprogs          3.x.0j                  # reiserfsck 2>&1|grep reiserfsprogs
 o  pcmcia-cs              3.1.21                  # cardmgr -V
 o  PPP                    2.4.0                   # pppd --version
