diff -urN parted-1.6.3.orig/libparted/disk_gpt.c parted-1.6.3/libparted/disk_gpt.c
--- parted-1.6.3.orig/libparted/disk_gpt.c	2002-04-12 15:43:14.000000000 -0600
+++ parted-1.6.3/libparted/disk_gpt.c	2003-06-08 18:32:52.000000000 -0600
@@ -51,7 +51,7 @@
 #define EFI_PMBR_OSTYPE_EFI 0xEE
 #define MSDOS_MBR_SIGNATURE 0xaa55
 
-#define GPT_HEADER_SIGNATURE 0x5452415020494645
+#define GPT_HEADER_SIGNATURE 0x5452415020494645ULL
 #define GPT_HEADER_REVISION_V1_02 0x00010200
 #define GPT_HEADER_REVISION_V1_00 0x00010000
 #define GPT_HEADER_REVISION_V0_99 0x00009900
diff -urN parted-1.6.3.orig/libparted/fs_ext2/ext2.c parted-1.6.3/libparted/fs_ext2/ext2.c
--- parted-1.6.3.orig/libparted/fs_ext2/ext2.c	2002-04-07 16:59:45.000000000 -0600
+++ parted-1.6.3/libparted/fs_ext2/ext2.c	2003-06-08 18:32:33.000000000 -0600
@@ -355,8 +355,8 @@
 	inode->i_blocks = PED_CPU_TO_LE32(EXT2_INODE_BLOCKS(*inode)
 		       			  + delta * i512perblock);
 	size = EXT2_INODE_SIZE(*inode) + delta * fs->blocksize;
-	inode->i_size = PED_CPU_TO_LE32(size % 0x100000000);
-	inode->i_size_high = PED_CPU_TO_LE32(size / 0x100000000);
+        inode->i_size = PED_CPU_TO_LE32(size % 0x100000000LL);
+        inode->i_size_high = PED_CPU_TO_LE32(size / 0x100000000LL);
 	inode->i_mtime = PED_CPU_TO_LE32(time(NULL));
 }
 
