--- reiserfsprogs-3.6.19/include/reiserfs_fs.h	2004-09-16 09:49:05.000000000 +0200
+++ reiserfsprogs-3.6.19-systemimager/include/reiserfs_fs.h	2007-08-02 16:33:50.000000000 +0200
@@ -38,14 +38,9 @@
 # define extern_inline
 #endif
 
-#include <asm/unaligned.h>
-
 #ifndef get_unaligned
-#if defined(__ppc__) || defined(ppc) || defined(__ppc) || \
-    defined(__PPC) || defined(powerpc) || defined(__powerpc__)
-#	define get_unaligned(ptr)	(*(ptr))
-#	define put_unaligned(val,ptr)	((void)(*(ptr) = (val)))
-#endif
+#define get_unaligned(ptr)	(*(ptr))
+#define put_unaligned(val,ptr)	((void)(*(ptr) = (val)))
 #endif
 
 #define get_leXX(xx,p,field)	(le##xx##_to_cpu ((p)->field))
