--- tar-1.13.25.orig/debian/README.Debian
+++ tar-1.13.25/debian/README.Debian
@@ -0,0 +1,18 @@
+The options -Z, --compress, and --uncompress all require that there be a
+'compress' utility installed on the system.  If there is no compress installed
+on your system, you will get errors if you try to use these options to tar.  
+
+Unfortunately, the algorithm used by compress is encumbered, and therefore 
+this utility is not part of Debian.
+
+The gunzip package can uncompress a file created with compress, but gzip can
+not create a file in 'compress' format.  Since the tar -Z option and friends
+end up calling 'compress' or 'compress -d', support for these options is not
+available in Debian.  However, at the time of this writing a package called 
+'ncompress' exists in non-free, and you can install it if the license terms 
+are acceptable to you.  
+
+Alternatively, if you don't really care what the format of the resulting file
+is, you could always create a 'compress' that is a symlink to gzip.  I refuse
+to do that in either the tar or gzip packages, however, since what gzip creates
+is a compressed file, but not a 'compress' format file...
--- tar-1.13.25.orig/debian/changelog
+++ tar-1.13.25/debian/changelog
@@ -0,0 +1,331 @@
+tar (1.13.25-2) unstable; urgency=medium
+
+  * add a README.Debian that clarifies the situation with respect to 'compress'
+    in Debian and the impact on the -Z and related options, closes: #122336
+  * patch from Mark Eichin to fix archive corruption in special cases, which
+    has been accepted upstream for next release.  closes: #126274
+
+ -- Bdale Garbee <bdale@gag.com>  Fri,  1 Feb 2002 08:49:26 -0700
+
+tar (1.13.25-1) unstable; urgency=medium
+
+  * new upstream version (bug fixes), closes: #113531
+  * start having tar provide rmt, which means conflicting with and replacing
+    cpio versions prior to the cutover, closes: #94257, #90794
+  * make medium urgency, since we really want this and the associated cpio
+    upload to both be in woody!
+
+ -- Bdale Garbee <bdale@gag.com>  Fri, 23 Nov 2001 09:20:40 -0700
+
+tar (1.13.22-1) unstable; urgency=medium
+
+  * new upstream version, released specifically to help close bugs in woody
+    upstream (Paul Eggert) says:
+      regarding 1.13.22
+        This fixes Debian bug 92106, in addition to the bug fixes I already
+        reported to you for GNU tar 1.13.20 and 1.13.21.  It also fixes a
+        core-dump bug for tar 1.13.19 and later, reported to bug-tar.
+      regarding 1.13.21
+        This fixes Debian bug 95984, in addition to the bug fixes I already
+        reported to you for 1.13.20.  It also upgrades tar to use gettext
+        0.10.39.
+      regarding 1.13.20
+        I haven't had time to fix all or even most of the bugs, but I suggest
+        closing out or modifying the following bug reports:
+        13312 I changed tar to avoid the problem (I hope; I can't test it).
+        52092 Fixed.
+        58890 Fixed, I think -- at least, I can't reproduce it now.
+        65719 Not a bug? last message in that bug report says it works for him.
+        77664 Not a bug.  In that context FOO:BAR means 'file BAR on host FOO'.
+        78179 Sorry, I don't follow this report.  Tar does strip leading '/'s 
+        	for me.
+        83458 Fixed.
+        83735 Fixed.
+        85400 Fixed for the info page only.  The man page is not part of 
+        	tar-1.13.20.
+        90794 This partly seems to be a Debian packaging problem; see 94257.
+        94287 Fixed.
+        95344 Fixed.
+        95984 Not fixed in 1.13.20, but will be fixed in next version.
+        99655 Fixed.
+        100883 This bug report applies to the Debian distribution only.
+        100885 Not a bug; see its last message.
+        105744 Not a bug; see its last message.
+    closes: #92106, #95984, #13312, #52092, #58890, #65719, #77664, #78179
+    closes: #83458, #83735, #94287, #95344, #99655, #100885, #105744
+  * add documentation for --rsh-command to the Debian-provided man page,
+    closes: #85400
+  * fix Debian-provided man page's reference to --exclude, closes: #100883
+
+ -- Bdale Garbee <bdale@gag.com>  Fri, 31 Aug 2001 01:13:49 -0600
+
+tar (1.13.19-1) unstable; urgency=low
+
+  * new upstream version, -I no longer a valid option, closes: #81556
+  * freshen debian/copyright
+
+ -- Bdale Garbee <bdale@gag.com>  Tue, 23 Jan 2001 09:42:13 -0700
+
+tar (1.13.18-2) unstable; urgency=low
+
+  * update (Debian-only, not part of upstream release) man page for tar to
+    reflect change from -I to -j for bzip2 support, closes: #80331
+
+ -- Bdale Garbee <bdale@gag.com>  Fri, 22 Dec 2000 14:18:51 -0700
+
+tar (1.13.18-1) unstable; urgency=low
+
+  * new upstream version, closes: #57436, #51889
+  * add suggestion for bzip2, closes: #64279
+  * this package is pristine upstream source plus the debian/ directory, so
+    there should be no issues compiling on any platform, closes: #58171
+
+ -- Bdale Garbee <bdale@gag.com>  Wed, 20 Dec 2000 13:09:44 -0700
+
+tar (1.13.17-2) frozen unstable; urgency=low
+
+  * reconfigure, recompile to fix compile from source problem, closes: #60824
+
+ -- Bdale Garbee <bdale@gag.com>  Thu, 23 Mar 2000 09:27:16 -0700
+
+tar (1.13.17-1) unstable; urgency=low
+
+  * new upstream source from alpha.gnu.org recommended by uptream maintainer
+    Paul Eggert.  
+  * this version should handle multibyte encoded filenames, closes: #25140
+  * upstream says the problem reported with -g is unreproducible in this
+    version, closes: #23511
+  * this version excludes sockets when building archives, closes: #51064
+
+ -- Bdale Garbee <bdale@gag.com>  Fri,  7 Jan 2000 22:57:34 -0700
+
+tar (1.13.15-1) unstable; urgency=low
+
+  * new upstream source from alpha.gnu.org recommended by uptream maintainer
+    Paul Eggert.
+  * update to current policy
+  * can't reproduce problem with remote host access reported in 1.13.11-2,
+    assuming it's fixed, closes: #45647
+  * upstream has not picked up our tar.1 manpage, so we'll try to keep it
+    up to date, closes: #50856
+
+ -- Bdale Garbee <bdale@gag.com>  Fri,  3 Dec 1999 13:29:41 -0700
+
+tar (1.13.14-5) unstable; urgency=low
+
+  * minor tweaks to clean up our diff, pointed out by the upstream maintainer
+
+ -- Bdale Garbee <bdale@gag.com>  Tue, 23 Nov 1999 09:23:56 -0700
+
+tar (1.13.14-4) unstable; urgency=low
+
+  * update upstream maintainer and copyright references, etc
+
+ -- Bdale Garbee <bdale@gag.com>  Sun, 21 Nov 1999 09:56:26 -0700
+
+tar (1.13.14-3) unstable; urgency=low
+
+  * fix default device in man page, closes: 50856
+
+ -- Bdale Garbee <bdale@gag.com>  Sun, 21 Nov 1999 09:45:32 -0700
+
+tar (1.13.14-2) unstable; urgency=low
+
+  * fold in Torsten's work that closes: #50553
+  * upstream folks indicate that $TAPE does not override -f in 1.13.14,
+    closes: #47664
+  * 1.13.14 has an updated man page that closes: #48603
+  * the -X stuff has supposedly been fixed since 1.13.12, closes: #43826
+  * rename upstream ChangeLog to changelog in the Debian package to satisfy
+    lintian/policy
+
+ -- Bdale Garbee <bdale@gag.com>  Sat, 20 Nov 1999 22:07:46 -0700
+
+tar (1.13.14-1.1) unstable; urgency=low
+
+  * Non maintainer upload.
+  * Moved the AC_LINK_FILES in configure.in inside the if (closes: #50553)
+
+ -- Torsten Landschoff <torsten@debian.org>  Sun, 21 Nov 1999 04:25:10 +0100
+
+tar (1.13.14-1) unstable; urgency=low
+
+  * new upstream source from alpha.gnu.org recommended by uptream maintainer
+    Paul Eggert.
+
+ -- Bdale Garbee <bdale@gag.com>  Thu, 11 Nov 1999 01:56:54 -0700
+
+tar (1.13.11-2) unstable; urgency=low
+
+  * fix error in man page, closes: #44610
+  * patch from upstream that closes: #44827
+
+ -- Bdale Garbee <bdale@gag.com>  Tue, 14 Sep 1999 20:58:23 -0600
+
+tar (1.13.11-1) unstable; urgency=low
+
+  * new upstream source from alpha.gnu.org recommended by uptream maintainer
+    Paul Eggert.
+  * move to FHS compliance with new debhelper
+
+ -- Bdale Garbee <bdale@gag.com>  Mon,  6 Sep 1999 16:49:51 -0600
+
+tar (1.13.6-1) unstable; urgency=medium
+
+  * new upstream source from alpha.gnu.org recommended by uptream maintainer
+    Paul Eggert.
+  * bzip2 support is back, now with option '-y' instead of '-I'... rejoice!
+    Closes: #42428, #42562, #42572, #42661, #42772
+  * Paul claims that the -X option is fixed again, Closes: #42552
+  * add mime-support goo, Closes: #26706
+  * close bug reports I forgot to close last time, Closes: #33134, #37659
+
+ -- Bdale Garbee <bdale@gag.com>  Wed, 11 Aug 1999 09:20:48 -0600
+
+tar (1.13-3) unstable; urgency=low
+
+  * include more docs in /usr/doc/tar, closes 33134
+  * minor tweak to tar.1, closes 37659
+
+ -- Bdale Garbee <bdale@gag.com>  Mon,  2 Aug 1999 00:48:44 -0600
+
+tar (1.13-2) unstable; urgency=low
+
+  * back out hacks we had made to 1.12 that seem to be causing problems in
+    1.13, getting essentially to pristine 1.13 source plus debian control files
+
+ -- Bdale Garbee <bdale@gag.com>  Mon,  2 Aug 1999 00:26:49 -0600
+
+tar (1.13-1) unstable; urgency=low
+
+  * new upstream source
+
+ -- Bdale Garbee <bdale@gag.com>  Sun,  1 Aug 1999 23:33:29 -0600
+
+tar (1.12-9) unstable; urgency=low
+
+  * fix some issues with the contest of the po directory in my CVS repository
+
+ -- Bdale Garbee <bdale@gag.com>  Sun,  1 Aug 1999 18:02:35 -0600
+
+tar (1.12-8) unstable; urgency=low
+
+  * update to handle changes in automake
+
+ -- Bdale Garbee <bdale@gag.com>  Wed,  9 Jun 1999 20:15:15 -0600
+
+tar (1.12-7) frozen unstable; urgency=low
+
+  * merge new version of Italian translation, closes 30284
+  * merge README.debian with copyright, closes 22370
+
+ -- Bdale Garbee <bdale@gag.com>  Wed,  6 Jan 1999 23:56:47 -0700
+
+tar (1.12-6) frozen unstable; urgency=low
+
+  * update manpage to document -I, closes 21224
+
+ -- Bdale Garbee <bdale@gag.com>  Fri, 17 Apr 1998 23:37:02 -0600
+
+tar (1.12-5) frozen unstable; urgency=low
+
+  * fix for erroneous time reports from --totals from Rob Browning, 
+    closes 18345
+  * add --numeric-owner to man page, closes 20801
+  * add some examples to the man page, closes 20290
+
+ -- Bdale Garbee <bdale@gag.com>  Fri, 10 Apr 1998 00:30:17 -0600
+
+tar (1.12-4) unstable; urgency=low
+
+  * move from debmake to debhelper
+  * address lintian error reports
+  * apply patch from amanda distribution to fix read errors on sparse files.
+    This should close 16694.
+  * updated dds2tar patch to restore proper operation of 'v' option.  This was
+    causing corrupted archives when stdout was used.  Closes 17857, 17916.
+
+ -- Bdale Garbee <bdale@gag.com>  Mon,  9 Feb 1998 23:52:05 -0700
+
+tar (1.12-3) unstable; urgency=low
+
+  * apply patch to support dds2tar-2.4.15, closes bug 10774
+  * apply patch to support use of bzip2, closes bugs 16280, 17221
+
+ -- Bdale Garbee <bdale@gag.com>  Wed,  4 Feb 1998 04:05:43 -0700
+
+tar (1.12-2) unstable; urgency=low
+
+  * libc6
+  * Folded in some of Michael Dorman's changes for alpha, which are really
+    libc6 changes.  Closes bug 8823.
+
+ -- Bdale Garbee <bdale@gag.com>  Fri,  5 Sep 1997 00:13:00 -0600
+
+tar (1.12-1) unstable; urgency=low
+
+  * New upstream version.
+
+ -- Bdale Garbee <bdale@gag.com>  Fri, 25 Apr 1997 21:15:44 -0600
+
+tar (1.11.8-11) stable frozen unstable; urgency=low
+
+  * back out the change made for 1.11.8-8, since it isn't really effective,
+    and caused several folks grief.  Closes bug 8040.
+
+ -- Bdale Garbee <bdale@gag.com>  Tue, 18 Mar 1997 01:38:10 -0700
+
+tar (1.11.8-10) stable frozen unstable; urgency=low
+
+  * rework debian/rules for CFLAGS as per policy.  Closes bug 8065.
+
+ -- Bdale Garbee <bdale@gag.com>  Fri, 14 Mar 1997 21:22:50 -0700
+
+tar (1.11.8-9) unstable; urgency=medium
+
+  * debmake shoved a man page for a porting utility (ansi2knr) that
+    is in the tar source tree into the package.  Fixes bug 7408.
+
+ -- Bdale Garbee <bdale@gag.com>  Thu, 27 Feb 1997 01:30:56 -0700
+
+tar (1.11.8-8) unstable; urgency=medium
+
+  * patch from the net for a quasi-security issue.  changes the behavior
+    during extracts, so that tar won't create inappropriate setuid files
+    from nonexistent users.
+
+ -- Bdale Garbee <bdale@gag.com>  Thu, 6 Feb 1997 23:06:17 -0700
+
+tar (1.11.8-7) unstable; urgency=medium
+
+  * updated to current package standards
+  * patch from the net that fixes sporadic multi-vol seg faults
+
+ -- Bdale Garbee <bdale@gag.com>  Sun, 17 Nov 1996 19:37:43 -0700
+
+Fri May 24 08:05:55 MDT 1996	Bdale Garbee	<bdale@gag.com>
+
+  * fold in changes from 1.11.9 configure.in so that we properly
+    recognize that it's ok to build 'rmt'
+  * administrivia
+
+Sun Apr 14 21:50:21 MDT 1996	Bdale Garbee	<bdale@gag.com>
+
+  * add architecture and section fields to the control file
+  * add a man page to squelch the frequent bug reports
+
+Wed Dec 27 00:29:37 MST 1995	Bdale Garbee	<bdale@gag.com>
+
+  * fixed a null pointer dereference when the username on a remote
+    tape reference was omitted... this closes several essentially 
+    duplicate bug reports about segmentation violation core dumps.
+
+Sun Dec  3 01:13:19 MST 1995	Bdale Garbee	<bdale@gag.com>
+
+  * ELF build
+  * add libc5 dependency
+  * new maintainer
+
+Local variables:
+mode: debian-changelog
+End:
--- tar-1.13.25.orig/debian/conffiles
+++ tar-1.13.25/debian/conffiles
@@ -0,0 +1 @@
+/etc/rmt
--- tar-1.13.25.orig/debian/control
+++ tar-1.13.25/debian/control
@@ -0,0 +1,20 @@
+Source: tar
+Section: base
+Priority: required
+Maintainer: Bdale Garbee <bdale@gag.com>
+Build-Depends: autoconf, automake, debhelper, gettext, rsh-client
+Standards-Version: 3.5.6.0
+
+Package: tar
+Architecture: any
+Pre-Depends: ${shlibs:Depends}
+Essential: yes
+Conflicts: cpio (<= 2.4.2-38)
+Replaces: cpio
+Suggests: bzip2
+Description: GNU tar
+ Tar is a program for packaging a set of files as a single archive in tar
+ format.  The function it performs is conceptually similar to cpio, and to
+ things like pkzip in the DOS world.  It is heavily used by the Debian package
+ management system, and is useful for performing system backups and exchanging
+ sets of files with others.
--- tar-1.13.25.orig/debian/copyright
+++ tar-1.13.25/debian/copyright
@@ -0,0 +1,20 @@
+This is the Debian GNU/Linux prepackaged version of GNU tar.  GNU
+tar, heavily based on John Gilmore's public domain version of tar,
+was originally written by Graham Todd.  It is now maintained by
+Paul Eggert <eggert@twinsun.com>.
+
+This package is maintained for Debian by Bdale Garbee <bdale@gag.com>, and
+was built from the sources found at:
+
+	ftp://alpha.gnu.org/gnu/tar/
+
+GNU tar is 
+
+Copyright 1994, 95, 96, 97, 98, 99, 2000, 2001 Free Software Foundation, Inc.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 dated June, 1991.
+
+On Debian GNU/Linux systems, the complete text of the GNU General Public 
+License can be found in the `/usr/share/common-licenses' directory.
--- tar-1.13.25.orig/debian/dirs
+++ tar-1.13.25/debian/dirs
@@ -0,0 +1,3 @@
+etc
+sbin
+usr/share/man/man8
--- tar-1.13.25.orig/debian/docs
+++ tar-1.13.25/debian/docs
@@ -0,0 +1,6 @@
+AUTHORS 
+ChangeLog
+ChangeLog.1
+NEWS 
+README
+THANKS
--- tar-1.13.25.orig/debian/mime
+++ tar-1.13.25/debian/mime
@@ -0,0 +1,2 @@
+application/x-tar; /bin/tar tvf -; print=/bin/tar tvf - | print text/plain:-; copiousoutput
+application/x-gtar; /bin/tar tvzf -; print=/bin/tar tvzf - | print text/plain:-; copiousoutput
--- tar-1.13.25.orig/debian/postinst
+++ tar-1.13.25/debian/postinst
@@ -0,0 +1,8 @@
+#!/bin/sh -e
+
+if [ "$1" = configure ]; then
+	install-info --quiet --section "General Commands" "General Commands" \
+		/usr/share/info/tar.info.gz
+fi
+
+#DEBHELPER#
--- tar-1.13.25.orig/debian/preinst
+++ tar-1.13.25/debian/preinst
@@ -0,0 +1,7 @@
+#!/bin/sh
+
+set -e
+
+dpkg --assert-support-predepends
+
+#DEBHELPER#
--- tar-1.13.25.orig/debian/prerm
+++ tar-1.13.25/debian/prerm
@@ -0,0 +1,10 @@
+#!/bin/sh -e
+
+if [ "$1" = remove ]; then 
+	install-info --quiet --remove tar
+	if test -L /sbin/rmt -a /sbin/rmt -ef /usr/sbin/rmt; then
+		rm -f /sbin/rmt
+	fi
+fi
+
+#DEBHELPER#
--- tar-1.13.25.orig/debian/rmt.8
+++ tar-1.13.25/debian/rmt.8
@@ -0,0 +1,223 @@
+.\" Copyright (c) 1983, 1991, 1993
+.\"	The Regents of the University of California.  All rights reserved.
+.\"
+.\" Redistribution and use in source and binary forms, with or without
+.\" modification, are permitted provided that the following conditions
+.\" are met:
+.\" 1. Redistributions of source code must retain the above copyright
+.\"    notice, this list of conditions and the following disclaimer.
+.\" 2. Redistributions in binary form must reproduce the above copyright
+.\"    notice, this list of conditions and the following disclaimer in the
+.\"    documentation and/or other materials provided with the distribution.
+.\" 3. All advertising materials mentioning features or use of this software
+.\"    must display the following acknowledgement:
+.\"	This product includes software developed by the University of
+.\"	California, Berkeley and its contributors.
+.\" 4. Neither the name of the University nor the names of its contributors
+.\"    may be used to endorse or promote products derived from this software
+.\"    without specific prior written permission.
+.\"
+.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
+.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+.\" SUCH DAMAGE.
+.\"
+.\"     @(#)rmt.8	8.2 (Berkeley) 12/11/93
+.\"
+.Dd December 11, 1993
+.Dt RMT 8
+.Os BSD 4.2
+.Sh NAME
+.Nm rmt
+.Nd remote magtape protocol module
+.Sh SYNOPSIS
+.Nm rmt
+.Sh DESCRIPTION
+.Nm Rmt
+is a program used by tar, cpio, mt, and the remote dump and restore
+programs in manipulating a magnetic tape drive through an interprocess
+communication connection.
+.Nm Rmt
+is normally started up with an
+.Xr rexec 3
+or
+.Xr rcmd 3
+call or the
+.Xr rsh 1
+command.
+.Pp
+The 
+.Nm rmt
+program accepts requests specific to the manipulation of
+magnetic tapes, performs the commands, then responds with
+a status indication.  All responses are in
+.Tn ASCII
+and in
+one of two forms. 
+Successful commands have responses of:
+.Bd -filled -offset indent
+.Sm off
+.Sy A Ar number No \en
+.Sm on
+.Ed
+.Pp
+.Ar Number
+is an
+.Tn ASCII
+representation of a decimal number.
+Unsuccessful commands are responded to with:
+.Bd -filled -offset indent
+.Sm off
+.Xo Sy E Ar error-number
+.No \en Ar error-message
+.No \en
+.Xc
+.Sm on
+.Ed
+.Pp
+.Ar Error-number
+is one of the possible error
+numbers described in
+.Xr intro 2
+and
+.Ar error-message
+is the corresponding error string as printed
+from a call to
+.Xr perror 3 .
+The protocol is comprised of the
+following commands, which are sent as indicated - no spaces are supplied
+between the command and its arguments, or between its arguments, and
+.Ql \en
+indicates that a newline should be supplied:
+.Bl -tag -width Ds
+.Sm off
+.It Xo Sy \&O Ar device
+.No \en Ar mode No \en
+.Xc
+Open the specified 
+.Ar device
+using the indicated
+.Ar mode .
+.Ar Device
+is a full pathname and
+.Ar mode
+is an
+.Tn ASCII
+representation of a decimal
+number suitable for passing to
+.Xr open 2 .
+If a device had already been opened, it is
+closed before a new open is performed.
+.It Xo Sy C Ar device No \en
+.Xc
+Close the currently open device.  The
+.Ar device
+specified is ignored.
+.It Xo Sy L
+.Ar offset No \en
+.Ar whence No \en
+.Xc
+.Sm on
+Perform an
+.Xr lseek 2
+operation using the specified parameters.
+The response value is that returned from the
+.Xr lseek
+call.
+.Sm off
+.It Sy W Ar count No \en
+.Sm on
+Write data onto the open device.
+.Nm Rmt
+reads
+.Ar count
+bytes from the connection, aborting if
+a premature end-of-file is encountered.
+The response value is that returned from
+the
+.Xr write 2
+call.
+.Sm off
+.It Sy R Ar count No \en
+.Sm on
+Read
+.Ar count
+bytes of data from the open device.
+If
+.Ar count
+exceeds the size of the data buffer (10 kilobytes), it is
+truncated to the data buffer size.
+.Nm rmt
+then performs the requested 
+.Xr read 2
+and responds with 
+.Sm off
+.Sy A Ar count-read No \en
+.Sm on
+if the read was
+successful; otherwise an error in the
+standard format is returned.  If the read
+was successful, the data read is then sent.
+.Sm off
+.It Xo Sy I Ar operation
+.No \en Ar count No \en
+.Xc
+.Sm on
+Perform a
+.Dv MTIOCOP
+.Xr ioctl 2
+command using the specified parameters.
+The parameters are interpreted as the
+.Tn ASCII
+representations of the decimal values
+to place in the 
+.Ar mt_op
+and
+.Ar mt_count
+fields of the structure used in the
+.Xr ioctl
+call.  The return value is the
+.Ar count
+parameter when the operation is successful.
+.It Sy S
+Return the status of the open device, as
+obtained with a
+.Dv MTIOCGET
+.Xr ioctl
+call.  If the operation was successful,
+an ``ack'' is sent with the size of the
+status buffer, then the status buffer is
+sent (in binary).
+.El
+.Sm on
+.Pp
+Any other command causes 
+.Nm rmt
+to exit.
+.Sh DIAGNOSTICS
+All responses are of the form described above.
+.Sh SEE ALSO
+.Xr tar 1 ,
+.Xr cpio 1 ,
+.Xr mt 1 ,
+.Xr rsh 1 ,
+.Xr rcmd 3 ,
+.Xr rexec 3 ,
+.Xr mtio 4 ,
+.Xr rdump 8 ,
+.Xr rrestore 8
+.Sh BUGS
+People should be discouraged from using this for a remote
+file access protocol.
+.Sh HISTORY
+The
+.Nm
+command appeared in
+.Bx 4.2 .
--- tar-1.13.25.orig/debian/rmt.sh
+++ tar-1.13.25/debian/rmt.sh
@@ -0,0 +1,8 @@
+#!/bin/sh
+#
+# This is not a mistake.  This shell script (/etc/rmt) has been provided
+# for compatibility with other Unix-like systems, some of which have
+# utilities that expect to find (and execute) rmt in the /etc directory
+# on remote systems.
+#
+exec /usr/sbin/rmt
--- tar-1.13.25.orig/debian/rules
+++ tar-1.13.25/debian/rules
@@ -0,0 +1,78 @@
+#!/usr/bin/make -f
+# Sample debian/rules that uses debhelper.
+# GNU copyright 1997 to 1999 by Joey Hess.
+
+# Uncomment this to turn on verbose mode.
+#export DH_VERBOSE=1
+
+# This is the debhelper compatability version to use.
+export DH_COMPAT=2
+
+configure: configure-stamp
+configure-stamp:
+	dh_testdir
+
+	CFLAGS="-O2 -g -Wall" ./configure --prefix=/usr \
+		--infodir=`pwd`/debian/tar/usr/share/info 
+
+	touch configure-stamp
+
+build: configure-stamp build-stamp
+build-stamp:
+	dh_testdir
+
+	$(MAKE)
+
+	touch build-stamp
+
+clean:
+	dh_testdir
+	dh_testroot
+	rm -f build-stamp configure-stamp
+
+	-$(MAKE) distclean
+
+	dh_clean
+
+install: build
+	dh_testdir
+	dh_testroot
+	dh_clean -k
+	dh_installdirs
+
+	make install bindir=`pwd`/debian/tar/bin prefix=`pwd`/debian/tar/usr \
+		libexecdir=`pwd`/debian/tar/usr/sbin 
+	install -m 755 debian/rmt.sh debian/tar/etc/rmt
+	install -m 644 debian/rmt.8 debian/tar/usr/share/man/man8/rmt.8
+	ln -s /usr/sbin/rmt debian/tar/sbin/rmt
+
+# Build architecture-independent files here.
+binary-indep: build install
+
+# Build architecture-dependent files here.
+binary-arch: build install
+	dh_testdir
+	dh_testroot
+	dh_installdocs
+	mv -f debian/tar/usr/share/doc/tar/ChangeLog \
+		debian/tar/usr/share/doc/tar/changelog
+	mv -f debian/tar/usr/share/doc/tar/ChangeLog.1 \
+		debian/tar/usr/share/doc/tar/changelog.1
+	dh_installexamples
+	dh_installmenu
+	dh_installcron
+	dh_installmanpages ansi2knr.1 ChangeLog.1
+	dh_installinfo
+	dh_installchangelogs #CHANGELOGS#
+	dh_link
+	dh_strip
+	dh_compress
+	dh_fixperms
+	dh_installdeb
+	dh_shlibdeps
+	dh_gencontrol
+	dh_md5sums
+	dh_builddeb
+
+binary: binary-indep binary-arch
+.PHONY: build clean binary-indep binary-arch binary install configure
--- tar-1.13.25.orig/debian/tar.1
+++ tar-1.13.25/debian/tar.1
@@ -0,0 +1,376 @@
+.\" @(#)tar.1 1.11.1 93/19/22 PJV;
+.TH TAR 1 "22 September 1993"
+.SH NAME
+tar \- The GNU version of the tar archiving utility
+.SH SYNOPSIS
+.B tar
+[
+.B \-
+]
+.B A --catenate --concatenate \||\| c --create \||\| d --diff --compare \||\| r --append \||\| t --list \||\| u --update \||\| x -extract --get
+[
+.B --atime-preserve
+]
+[
+.B -b, --block-size N
+]
+[
+.B -B, --read-full-blocks
+]
+[
+.B -C, --directory DIR
+]
+[
+.B --checkpoint	
+]
+[
+.B -f, --file [HOSTNAME:]F
+]
+[
+.B --force-local	
+]
+[
+.B -F, --info-script F --new-volume-script F
+]
+[
+.B -G, --incremental
+]
+[
+.B -g, --listed-incremental F
+]
+[
+.B -h, --dereference
+]
+[
+.B -i, --ignore-zeros
+]
+[
+.B -j, --bzip2
+]
+[
+.B --ignore-failed-read
+]
+[
+.B -k, --keep-old-files
+]
+[
+.B -K, --starting-file F
+]
+[
+.B -l, --one-file-system
+]
+[
+.B -L, --tape-length N
+]
+[
+.B -m, --modification-time
+]
+[
+.B -M, --multi-volume
+]
+[
+.B -N, --after-date DATE, --newer DATE
+]
+[
+.B -o, --old-archive, --portability
+]
+[
+.B -O, --to-stdout
+]
+[
+.B -p, --same-permissions, --preserve-permissions
+]
+[
+.B -P, --absolute-names
+]
+[
+.B --preserve	
+]
+[
+.B -R, --record-number
+]
+[
+.B --remove-files
+]
+[
+.B -s, --same-order, --preserve-order
+]
+[
+.B --same-owner
+]
+[
+.B --numeric-owner
+]
+[
+.B -S, --sparse
+]
+[
+.B -T, --files-from F
+]
+[
+.B --null	
+]
+[
+.B --totals	
+]
+[
+.B -v, --verbose
+]
+[
+.B -V, --label NAME
+]
+[
+.B --version	
+]
+[
+.B -w, --interactive, --confirmation
+]
+[
+.B -W, --verify	
+]
+[
+.B --exclude=FILE
+]
+[
+.B -X, --exclude-from FILE
+]
+[
+.B -Z, --compress, --uncompress
+]
+[
+.B -z, --gzip, --ungzip	
+]
+[
+.B --use-compress-program PROG
+]
+[
+.B --block-compress
+]
+[
+.B --rsh-command=CMD
+]
+[
+.B -[0-7][lmh]	
+]
+.TP
+.I filename1 [ filename2, ... filenameN ] 
+.TP
+.I directory1 [ directory2, ...directoryN ]
+.SH DESCRIPTION
+.LP
+This manual page documents the GNU version of
+.B tar
+, an archiving program designed to store and extract files from 
+an archive file known as a 
+.IR tarfile.
+A 
+.IR tarfile 
+may be made on a tape drive, however, it is also common
+to write a
+.IR tarfile
+to a normal file. 
+The first argument to 
+.B tar
+must be one of the options:
+.BR Acdrtux ,
+followed by any optional functions.
+The final arguments to 
+.B tar
+are the names of the files or directories which should be archived. The use
+of a directory name always implies that the subdirectories below should be
+included in the archive.
+.SH EXAMPLES
+.TP
+.B tar -xvvf foo.tar
+extract foo.tar
+.TP
+.B tar -xvvzf foo.tar.gz
+extract gzipped foo.tar.gz
+.TP
+.B tar -cvvf foo.tar foo/
+tar contents of folder foo in foo.tar
+.SH "FUNCTION LETTERS"
+.TP
+.B One of the following options must be used:
+.TP
+.B -A, --catenate, --concatenate
+append tar files to an archive
+.TP
+.B -c, --create	
+create a new archive
+.TP
+.B -d, --diff, --compare
+find differences between archive and file system
+.TP
+.B --delete		
+delete from the archive (not for use on mag tapes!)
+.TP
+.B -r, --append		
+append files to the end of an archive
+.TP
+.B -t, --list		
+list the contents of an archive
+.TP
+.B -u, --update		
+only append files that are newer than copy in archive
+.TP
+.B -x, --extract, --get		
+extract files from an archive
+.SH "OTHER OPTIONS"
+.TP
+.B --atime-preserve	
+don't change access times on dumped files
+.TP
+.B -b, --block-size N	
+block size of Nx512 bytes (default N=20)
+.TP
+.B -B, --read-full-blocks	
+reblock as we read (for reading 4.2BSD pipes)
+.TP 
+.B -C, --directory DIR	
+change to directory DIR
+.TP 
+.B --checkpoint		
+print directory names while reading the archive
+.TP
+.B -f, --file [HOSTNAME:]F	
+use archive file or device F (default "-", meaning stdin/stdout)
+.TP
+.B --force-local		
+archive file is local even if has a colon
+.TP 
+.B -F, --info-script F --new-volume-script F 
+run script at end of each tape (implies -M)
+.TP
+.B -G, --incremental	
+create/list/extract old GNU-format incremental backup
+.TP
+.B -g, --listed-incremental F 
+create/list/extract new GNU-format incremental backup
+.TP 
+.B -h, --dereference	
+don't dump symlinks; dump the files they point to
+.TP
+.B -i, --ignore-zeros	
+ignore blocks of zeros in archive (normally mean EOF)
+.TP
+.B -j, --bzip2
+filter archive through bzip2, use to decompress .bz2 files
+.TP
+.B --ignore-failed-read	
+don't exit with non-zero status on unreadable files
+.TP
+.B -k, --keep-old-files	
+keep existing files; don't overwrite them from archive
+.TP
+.B -K, --starting-file F	
+begin at file F in the archive
+.TP
+.B -l, --one-file-system	
+stay in local file system when creating an archive
+.TP
+.B -L, --tape-length N	
+change tapes after writing N*1024 bytes
+.TP
+.B -m, --modification-time	
+don't extract file modified time
+.TP
+.B -M, --multi-volume	
+create/list/extract multi-volume archive
+.TP
+.B -N, --after-date DATE, --newer DATE	
+only store files newer than DATE
+.TP
+.B -o, --old-archive, --portability	
+write a V7 format archive, rather than ANSI format
+.TP
+.B -O, --to-stdout		
+extract files to standard output
+.TP
+.B -p, --same-permissions, --preserve-permissions 
+extract all protection information
+.TP
+.B -P, --absolute-paths	
+don't strip leading `/'s from file names
+.TP
+.B --preserve		
+like -p -s
+.TP
+.B -R, --record-number	
+show record number within archive with each message
+.TP 
+.B --remove-files		
+remove files after adding them to the archive
+.TP
+.B -s, --same-order, --preserve-order	
+list of names to extract is sorted to match archive
+.TP
+.B --same-owner		
+create extracted files with the same ownership 
+.TP
+.B --numeric-owner
+always use numbers for user/group names
+.TP
+.B -S, --sparse		
+handle sparse files efficiently
+.TP 
+.B -T, --files-from F	
+get names to extract or create from file F
+.TP
+.B --null			
+-T reads null-terminated names, disable -C
+.TP
+.B --totals		
+print total bytes written with --create
+.TP
+.B -v, --verbose		
+verbosely list files processed
+.TP
+.B -V, --label NAME	
+create archive with volume name NAME
+.TP 
+.B --version		
+print tar program version number
+.TP
+.B -w, --interactive, --confirmation	
+ask for confirmation for every action
+.TP
+.B -W, --verify		
+attempt to verify the archive after writing it
+.TP
+.B --exclude=FILE		
+exclude file FILE
+.TP
+.B -X, --exclude-from FILE	
+exclude files listed in FILE
+.TP
+.B -Z, --compress, --uncompress      	
+filter the archive through compress
+.TP 
+.B -z, --gzip, --ungzip		
+filter the archive through gzip
+.TP
+.B --use-compress-program PROG
+filter the archive through PROG (which must accept -d)
+.TP
+.B --block-compress	
+block the output of compression program for tapes
+.TP
+.B --rsh-command=CMD
+Use remote COMMAND instead of `rsh'.  This option exists so that
+people who use something other than the standard `rsh' (e.g., a
+Kerberized `rsh') can access a remote device.
+.TP
+.B -[0-7][lmh]		
+specify drive and density
+.SH BUGS
+.LP
+The GNU folks, in general, abhor man pages, and create info documents instead.
+The maintainer of tar falls into this category.  This man page is neither
+complete, nor current, and was included in the Debian Linux packaging of tar
+entirely to reduce the frequency with which the lack of a man page gets
+reported as a bug in our defect tracking system.
+
+If you really want to understand tar, then you should run info and read the
+tar info pages, or use the info mode in emacs.
+
--- tar-1.13.25.orig/src/buffer.c
+++ tar-1.13.25/src/buffer.c
@@ -419,8 +419,6 @@
 	{
 	  size_t size = record_size - length;
 
-	  if (size < BLOCKSIZE)
-	    size = BLOCKSIZE;
 	  status = safe_read (STDIN_FILENO, cursor, size);
 	  if (status <= 0)
 	    break;
