--- util-linux-2.12q/configure	Mon Dec 20 15:20:35 2004
+++ util-linux-2.12q/configure	Thu Apr 21 23:54:25 2005
@@ -66,9 +66,9 @@
 
 DEFS=
 LIBS=
-compile='$CC $CFLAGS $DEFS conftest.c -o conftest $LIBS >/dev/null 2>&1'
-compile_with_warnings='$CC $CFLAGS $DEFS conftest.c -o conftest $LIBS 2>&1'
-static_compile='$CC -static $DEFS conftest.c -o conftest $LIBS >/dev/null 2>&1'
+compile='$CC $CFLAGS $DEFS conftest.c -o conftest $LDFLAGS $LIBS >/dev/null 2>&1'
+compile_with_warnings='$CC $CFLAGS $DEFS conftest.c -o conftest $LDFLAGS $LIBS 2>&1'
+static_compile='$CC -static $DEFS conftest.c -o conftest $LDFLAGS $LIBS >/dev/null 2>&1'
 eval COMPILE='"'$compile'"'
 export COMPILE
 
