diff -ur util-linux-2.12r.old/disk-utils/fsck.cramfs.c util-linux-2.12r/disk-utils/fsck.cramfs.c
--- util-linux-2.12r.old/disk-utils/fsck.cramfs.c	2004-12-11 15:53:16.000000000 +0100
+++ util-linux-2.12r/disk-utils/fsck.cramfs.c	2006-04-21 18:28:48.000000000 +0200
@@ -76,7 +76,7 @@
 
 #define PAD_SIZE 512
 
-#include <asm/page.h>
+#include <sys/user.h>		/* PAGE_SIZE */
 #ifdef PAGE_SIZE
 #define PAGE_CACHE_SIZE ((int) PAGE_SIZE)
 #elif defined __ia64__
