--- util-linux-2.12r/fdisk/llseek.c.orig	2006-11-01 19:19:09.000000000 +0100
+++ util-linux-2.12r/fdisk/llseek.c	2006-11-01 19:19:16.000000000 +0100
@@ -31,9 +31,18 @@
 
 #ifdef __NR__llseek
 
+#ifdef _syscall5
+
 static _syscall5(int,_llseek,unsigned int,fd,unsigned long,offset_high,
 		 unsigned long, offset_low,long long *,result,
 		 unsigned int, origin)
+#else
+static int _llseek (unsigned int fd, unsigned long oh,
+                   unsigned long ol, long long *result,
+                   unsigned int origin) {
+       return syscall (__NR__llseek, fd, oh, ol, result, origin);
+}
+#endif /* _syscall5 */
 
 #else
 
--- util-linux-2.12r/fdisk/sfdisk.c.orig	2006-11-01 19:19:14.000000000 +0100
+++ util-linux-2.12r/fdisk/sfdisk.c	2006-11-01 19:19:20.000000000 +0100
@@ -177,9 +177,16 @@
 #endif
 
 #ifndef use_lseek
+#ifdef _syscall5
 static __attribute__used
 _syscall5(int,  _llseek,  unsigned int,  fd, ulong, hi, ulong, lo,
        loff_t *, res, unsigned int, wh);
+#else
+int _llseek (unsigned int fd, ulong hi, ulong lo,
+       loff_t *res, unsigned int wh) {
+       return syscall (__NR__llseek, fd, hi, lo, res, wh);
+}
+#endif /* _syscall5 */
 #endif
 
 static int
