--- sysvinit-2.88dsf/debian/rules.orig	2013-01-30 09:35:10.377190024 +0100
+++ sysvinit-2.88dsf/debian/rules	2013-01-30 09:37:30.465124757 +0100
@@ -18,32 +18,32 @@
 
 LC_ALL	= POSIX
 
-DEB_HOST_GNU_TYPE   ?= $(shell dpkg-architecture -qDEB_HOST_GNU_TYPE)
-DEB_HOST_GNU_SYSTEM ?= $(shell dpkg-architecture -qDEB_HOST_GNU_SYSTEM)
-DEB_HOST_ARCH       ?= $(shell dpkg-architecture -qDEB_HOST_ARCH)
-DEB_HOST_ARCH_OS    ?= $(shell dpkg-architecture -qDEB_HOST_ARCH_OS)
-DEB_HOST_MULTIARCH  ?= $(shell dpkg-architecture -qDEB_HOST_MULTIARCH)
+#DEB_HOST_GNU_TYPE   ?= $(shell dpkg-architecture -qDEB_HOST_GNU_TYPE)
+#DEB_HOST_GNU_SYSTEM ?= $(shell dpkg-architecture -qDEB_HOST_GNU_SYSTEM)
+#DEB_HOST_ARCH       ?= $(shell dpkg-architecture -qDEB_HOST_ARCH)
+#DEB_HOST_ARCH_OS    ?= $(shell dpkg-architecture -qDEB_HOST_ARCH_OS)
+#DEB_HOST_MULTIARCH  ?= $(shell dpkg-architecture -qDEB_HOST_MULTIARCH)
 
 INSTALL_DATA = install -o root -g root -m 644
 INSTALL      = install -o root -g root -m 755
 
 # Handle cross builds
-ifneq ($(DEB_HOST_GNU_TYPE),$(DEB_BUILD_GNU_TYPE))
-CROSS=CC=$(DEB_HOST_GNU_TYPE)-gcc
+#ifneq ($(DEB_HOST_GNU_TYPE),$(DEB_BUILD_GNU_TYPE))
+#CROSS=CC=$(DEB_HOST_GNU_TYPE)-gcc
 # CROSS is passed to make, CC is called directly.
-CC=$(DEB_HOST_GNU_TYPE)-gcc
-endif
+#CC=$(DEB_HOST_GNU_TYPE)-gcc
+#endif
 
 # Set arch specific build flags
-ifeq ($(DEB_HOST_ARCH_OS),linux)
-CONFFLAGS = WITH_SELINUX="yes"
-endif
+#ifeq ($(DEB_HOST_ARCH_OS),linux)
+#CONFFLAGS = WITH_SELINUX="yes"
+#endif
 
 PATCH_STAMP=debian/stamp-patched
 patch: $(PATCH_STAMP)
 $(PATCH_STAMP):
-	dh_testdir debian/patches/series \
-		debian/startpar/patches/series
+#	dh_testdir debian/patches/series \
+#		debian/startpar/patches/series
 	QUILT_PATCHES=debian/patches quilt --quiltrc /dev/null push -a || test $$? = 2
 	cd debian/startpar && \
 		QUILT_PATCHES=patches quilt --quiltrc /dev/null push -a || test $$? = 2
@@ -58,7 +58,7 @@
 build: build-stamp
 build-stamp: patch
 # Builds the binary package.
-	dh_testdir
+#	dh_testdir
 	$(MAKE) $(CROSS) $(CONFFLAGS) -C src DISTRO=Debian LIBDIR=/usr/lib/$(DEB_HOST_MULTIARCH)
 	$(MAKE) $(CROSS) -C debian/startpar
 	touch $@
@@ -140,11 +140,11 @@
 	dh_builddeb
 
 clean:
-	dh_testdir
+#	dh_testdir
 	$(MAKE) -C src clobber
 	$(MAKE) -C debian/startpar clean
 	$(MAKE) -f debian/rules unpatch
-	dh_clean build-stamp debian/copyright
+#	dh_clean build-stamp debian/copyright
 
 binary: binary-arch binary-indep
 
